// channelkit/dist/cli/commands/service-install.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.installServiceCommand = installServiceCommand;
exports.uninstallServiceCommand = uninstallServiceCommand;
exports.serviceStatusCommand = serviceStatusCommand;
exports.serviceStopCommand = serviceStopCommand;
exports.serviceStartCommand = serviceStartCommand;
exports.promptServiceInstall = promptServiceInstall;
const os_1 = require("os");
const path_1 = require("path");
const fs_1 = require("fs");
const child_process_1 = require("child_process");
const readline_1 = require("readline");
const helpers_1 = require("../helpers");
const paths_1 = require("../../paths");
const SERVICE_NAME = 'com.channelkit.server';
const SYSTEMD_SERVICE = 'channelkit';
function getNodePath() {
    try {
        return (0, child_process_1.execSync)('which node', { encoding: 'utf-8' }).trim();
    }
    catch {
        return 'node';
    }
}
function getChannelKitBin() {
    try {
        return (0, child_process_1.execSync)('which channelkit', { encoding: 'utf-8' }).trim();
    }
    catch {
        // Fallback: resolve from current execution
        return (0, path_1.resolve)((0, path_1.join)(__dirname, '..', '..', '..', 'dist', 'cli.js'));
    }
}
function getNpmGlobalPath() {
    try {
        return (0, child_process_1.execSync)('npm prefix -g', { encoding: 'utf-8' }).trim();
    }
    catch {
        return '';
    }
}
// ─── macOS (launchd) ───
function getMacPlistPath() {
    return (0, path_1.join)((0, os_1.homedir)(), 'Library', 'LaunchAgents', `${SERVICE_NAME}.plist`);
}
function generatePlist(configPath) {
    const nodePath = getNodePath();
    const ckBin = getChannelKitBin();
    const npmGlobal = getNpmGlobalPath();
    const path = ['/usr/local/bin', '/usr/bin', '/bin', '/usr/sbin', '/sbin'];
    if (npmGlobal)
        path.unshift((0, path_1.join)(npmGlobal, 'bin'));
    const nodeDir = nodePath.replace(/\/node$/, '');
    if (!path.includes(nodeDir))
        path.unshift(nodeDir);
    return `<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>Label</key>
  <string>${SERVICE_NAME}</string>
  <key>ProgramArguments</key>
  <array>
    <string>${nodePath}</string>
    <string>${ckBin}</string>
    <string>start</string>
    <string>-c</string>
    <string>${configPath}</string>
  </array>
  <key>WorkingDirectory</key>
  <string>${paths_1.CHANNELKIT_HOME}</string>
  <key>EnvironmentVariables</key>
  <dict>
    <key>PATH</key>
    <string>${path.join(':')}</string>
  </dict>
  <key>RunAtLoad</key>
  <true/>
  <key>KeepAlive</key>
  <true/>
  <key>StandardOutPath</key>
  <string>${(0, path_1.join)(paths_1.CHANNELKIT_HOME, 'channelkit.log')}</string>
  <key>StandardErrorPath</key>
  <string>${(0, path_1.join)(paths_1.CHANNELKIT_HOME, 'channelkit.log')}</string>
</dict>
</plist>`;
}
function installMac(configPath) {
    const plistPath = getMacPlistPath();
    const plist = generatePlist(configPath);
    // Ensure LaunchAgents directory exists
    const agentsDir = (0, path_1.join)((0, os_1.homedir)(), 'Library', 'LaunchAgents');
    if (!(0, fs_1.existsSync)(agentsDir))
        (0, fs_1.mkdirSync)(agentsDir, { recursive: true });
    // Unload existing service if present
    try {
        (0, child_process_1.execSync)(`launchctl unload "${plistPath}" 2>/dev/null`, { encoding: 'utf-8' });
    }
    catch { }
    (0, fs_1.writeFileSync)(plistPath, plist);
    console.log((0, helpers_1.c)('dim', `  Written: ${plistPath}`));
    try {
        (0, child_process_1.execSync)(`launchctl load "${plistPath}"`, { encoding: 'utf-8' });
        console.log((0, helpers_1.c)('green', '\n  ✅ Service installed and started!'));
        console.log((0, helpers_1.c)('dim', `\n  ChannelKit will now start automatically on login.`));
        console.log((0, helpers_1.c)('dim', `  Logs: ${(0, path_1.join)(paths_1.CHANNELKIT_HOME, 'channelkit.log')}`));
        console.log((0, helpers_1.c)('dim', `\n  Manage with:`));
        console.log((0, helpers_1.c)('cyan', `    channelkit service status`));
        console.log((0, helpers_1.c)('cyan', `    channelkit service stop`));
        console.log((0, helpers_1.c)('cyan', `    channelkit service uninstall\n`));
    }
    catch (err) {
        console.error((0, helpers_1.c)('yellow', `\n  ❌ Failed to load service: ${err.message}`));
        console.log((0, helpers_1.c)('dim', `  Try manually: launchctl load "${plistPath}"\n`));
    }
}
function uninstallMac() {
    const plistPath = getMacPlistPath();
    if (!(0, fs_1.existsSync)(plistPath)) {
        console.log((0, helpers_1.c)('dim', '\n  No service installed.\n'));
        return;
    }
    try {
        (0, child_process_1.execSync)(`launchctl unload "${plistPath}" 2>/dev/null`, { encoding: 'utf-8' });
    }
    catch { }
    (0, fs_1.unlinkSync)(plistPath);
    console.log((0, helpers_1.c)('green', '\n  ✅ Service uninstalled.\n'));
}
function statusMac() {
    const plistPath = getMacPlistPath();
    if (!(0, fs_1.existsSync)(plistPath)) {
        console.log((0, helpers_1.c)('dim', '\n  No service installed.\n'));
        return;
    }
    try {
        const output = (0, child_process_1.execSync)(`launchctl list | grep ${SERVICE_NAME}`, { encoding: 'utf-8' }).trim();
        if (output) {
            const parts = output.split(/\s+/);
            const pid = parts[0];
            const exitCode = parts[1];
            if (pid && pid !== '-') {
                console.log((0, helpers_1.c)('green', `\n  ✅ Running (PID ${pid})`));
            }
            else {
                console.log((0, helpers_1.c)('yellow', `\n  ⏹  Stopped (last exit code: ${exitCode})`));
            }
        }
        else {
            console.log((0, helpers_1.c)('dim', '\n  Service loaded but not found in process list.'));
        }
    }
    catch {
        console.log((0, helpers_1.c)('yellow', '\n  ⏹  Not running'));
    }
    console.log((0, helpers_1.c)('dim', `  Logs: ${(0, path_1.join)(paths_1.CHANNELKIT_HOME, 'channelkit.log')}\n`));
}
function stopMac() {
    const plistPath = getMacPlistPath();
    if (!(0, fs_1.existsSync)(plistPath)) {
        console.log((0, helpers_1.c)('dim', '\n  No service installed.\n'));
        return;
    }
    try {
        (0, child_process_1.execSync)(`launchctl unload "${plistPath}"`, { encoding: 'utf-8' });
        (0, child_process_1.execSync)(`launchctl load "${plistPath}"`, { encoding: 'utf-8' });
        // Actually, to just stop without removing:
    }
    catch { }
    try {
        (0, child_process_1.execSync)(`launchctl stop ${SERVICE_NAME}`, { encoding: 'utf-8' });
        console.log((0, helpers_1.c)('green', '\n  ⏹  Service stopped.\n'));
    }
    catch {
        console.log((0, helpers_1.c)('yellow', '\n  Service may not be running.\n'));
    }
}
function startMac() {
    const plistPath = getMacPlistPath();
    if (!(0, fs_1.existsSync)(plistPath)) {
        console.log((0, helpers_1.c)('yellow', '\n  No service installed. Run: channelkit service install\n'));
        return;
    }
    try {
        (0, child_process_1.execSync)(`launchctl load "${plistPath}" 2>/dev/null`, { encoding: 'utf-8' });
    }
    catch { }
    try {
        (0, child_process_1.execSync)(`launchctl start ${SERVICE_NAME}`, { encoding: 'utf-8' });
        console.log((0, helpers_1.c)('green', '\n  ▶  Service started.\n'));
    }
    catch (err) {
        console.error((0, helpers_1.c)('yellow', `\n  ❌ Failed to start: ${err.message}\n`));
    }
}
// ─── Linux (systemd) ───
function getSystemdPath() {
    const dir = (0, path_1.join)((0, os_1.homedir)(), '.config', 'systemd', 'user');
    if (!(0, fs_1.existsSync)(dir))
        (0, fs_1.mkdirSync)(dir, { recursive: true });
    return (0, path_1.join)(dir, `${SYSTEMD_SERVICE}.service`);
}
function generateSystemdUnit(configPath) {
    const nodePath = getNodePath();
    const ckBin = getChannelKitBin();
    return `[Unit]
Description=ChannelKit Messaging Gateway
After=network.target

[Service]
Type=simple
ExecStart=${nodePath} ${ckBin} start -c ${configPath}
WorkingDirectory=${paths_1.CHANNELKIT_HOME}
Restart=on-failure
RestartSec=5
StandardOutput=journal
StandardError=journal

[Install]
WantedBy=default.target
`;
}
function installLinux(configPath) {
    const unitPath = getSystemdPath();
    const unit = generateSystemdUnit(configPath);
    (0, fs_1.writeFileSync)(unitPath, unit);
    console.log((0, helpers_1.c)('dim', `  Written: ${unitPath}`));
    try {
        (0, child_process_1.execSync)('systemctl --user daemon-reload', { encoding: 'utf-8' });
        (0, child_process_1.execSync)(`systemctl --user enable ${SYSTEMD_SERVICE}`, { encoding: 'utf-8' });
        (0, child_process_1.execSync)(`systemctl --user start ${SYSTEMD_SERVICE}`, { encoding: 'utf-8' });
        // Enable lingering so service starts on boot even without login
        (0, child_process_1.execSync)('loginctl enable-linger', { encoding: 'utf-8' });
        console.log((0, helpers_1.c)('green', '\n  ✅ Service installed and started!'));
        console.log((0, helpers_1.c)('dim', `\n  ChannelKit will now start automatically on boot.`));
        console.log((0, helpers_1.c)('dim', `  Logs: journalctl --user -u ${SYSTEMD_SERVICE} -f`));
        console.log((0, helpers_1.c)('dim', `\n  Manage with:`));
        console.log((0, helpers_1.c)('cyan', `    channelkit service status`));
        console.log((0, helpers_1.c)('cyan', `    channelkit service stop`));
        console.log((0, helpers_1.c)('cyan', `    channelkit service uninstall\n`));
    }
    catch (err) {
        console.error((0, helpers_1.c)('yellow', `\n  ❌ Failed to start service: ${err.message}`));
        console.log((0, helpers_1.c)('dim', `  Try manually: systemctl --user start ${SYSTEMD_SERVICE}\n`));
    }
}
function uninstallLinux() {
    const unitPath = getSystemdPath();
    if (!(0, fs_1.existsSync)(unitPath)) {
        console.log((0, helpers_1.c)('dim', '\n  No service installed.\n'));
        return;
    }
    try {
        (0, child_process_1.execSync)(`systemctl --user stop ${SYSTEMD_SERVICE} 2>/dev/null`, { encoding: 'utf-8' });
        (0, child_process_1.execSync)(`systemctl --user disable ${SYSTEMD_SERVICE} 2>/dev/null`, { encoding: 'utf-8' });
    }
    catch { }
    (0, fs_1.unlinkSync)(unitPath);
    try {
        (0, child_process_1.execSync)('systemctl --user daemon-reload', { encoding: 'utf-8' });
    }
    catch { }
    console.log((0, helpers_1.c)('green', '\n  ✅ Service uninstalled.\n'));
}
function statusLinux() {
    const unitPath = getSystemdPath();
    if (!(0, fs_1.existsSync)(unitPath)) {
        console.log((0, helpers_1.c)('dim', '\n  No service installed.\n'));
        return;
    }
    try {
        const output = (0, child_process_1.execSync)(`systemctl --user is-active ${SYSTEMD_SERVICE}`, { encoding: 'utf-8' }).trim();
        if (output === 'active') {
            console.log((0, helpers_1.c)('green', '\n  ✅ Running'));
        }
        else {
            console.log((0, helpers_1.c)('yellow', `\n  ⏹  ${output}`));
        }
    }
    catch {
        console.log((0, helpers_1.c)('yellow', '\n  ⏹  Not running'));
    }
    console.log((0, helpers_1.c)('dim', `  Logs: journalctl --user -u ${SYSTEMD_SERVICE} -f\n`));
}
function stopLinux() {
    try {
        (0, child_process_1.execSync)(`systemctl --user stop ${SYSTEMD_SERVICE}`, { encoding: 'utf-8' });
        console.log((0, helpers_1.c)('green', '\n  ⏹  Service stopped.\n'));
    }
    catch {
        console.log((0, helpers_1.c)('yellow', '\n  Service may not be running.\n'));
    }
}
function startLinux() {
    const unitPath = getSystemdPath();
    if (!(0, fs_1.existsSync)(unitPath)) {
        console.log((0, helpers_1.c)('yellow', '\n  No service installed. Run: channelkit service install\n'));
        return;
    }
    try {
        (0, child_process_1.execSync)(`systemctl --user start ${SYSTEMD_SERVICE}`, { encoding: 'utf-8' });
        console.log((0, helpers_1.c)('green', '\n  ▶  Service started.\n'));
    }
    catch (err) {
        console.error((0, helpers_1.c)('yellow', `\n  ❌ Failed to start: ${err.message}\n`));
    }
}
// ─── Public API ───
const os = (0, os_1.platform)();
function installServiceCommand(configPath) {
    const resolved = (0, path_1.resolve)(configPath);
    console.log((0, helpers_1.c)('cyan', '\n  📦 Installing ChannelKit as a system service...\n'));
    if (os === 'darwin') {
        installMac(resolved);
    }
    else if (os === 'linux') {
        installLinux(resolved);
    }
    else {
        console.log((0, helpers_1.c)('yellow', `\n  ❌ Unsupported platform: ${os}`));
        console.log((0, helpers_1.c)('dim', '  Supported: macOS (launchd), Linux (systemd)\n'));
    }
}
function uninstallServiceCommand() {
    if (os === 'darwin')
        uninstallMac();
    else if (os === 'linux')
        uninstallLinux();
    else
        console.log((0, helpers_1.c)('yellow', `\n  ❌ Unsupported platform: ${os}\n`));
}
function serviceStatusCommand() {
    if (os === 'darwin')
        statusMac();
    else if (os === 'linux')
        statusLinux();
    else
        console.log((0, helpers_1.c)('yellow', `\n  ❌ Unsupported platform: ${os}\n`));
}
function serviceStopCommand() {
    if (os === 'darwin')
        stopMac();
    else if (os === 'linux')
        stopLinux();
    else
        console.log((0, helpers_1.c)('yellow', `\n  ❌ Unsupported platform: ${os}\n`));
}
function serviceStartCommand() {
    if (os === 'darwin')
        startMac();
    else if (os === 'linux')
        startLinux();
    else
        console.log((0, helpers_1.c)('yellow', `\n  ❌ Unsupported platform: ${os}\n`));
}
/** Prompt user to install service after first successful start */
async function promptServiceInstall(configPath) {
    // Don't prompt if already installed
    if (os === 'darwin' && (0, fs_1.existsSync)(getMacPlistPath()))
        return;
    if (os === 'linux' && (0, fs_1.existsSync)(getSystemdPath()))
        return;
    if (os !== 'darwin' && os !== 'linux')
        return;
    const rl = (0, readline_1.createInterface)({ input: process.stdin, output: process.stdout });
    try {
        console.log();
        const answer = await (0, helpers_1.ask)(rl, 'Install as a system service? (starts on boot) [y/N]', 'N');
        if (answer.toLowerCase() === 'y') {
            rl.close();
            installServiceCommand(configPath);
        }
        else {
            console.log((0, helpers_1.c)('dim', '\n  No problem! You can install later with: channelkit service install\n'));
        }
    }
    finally {
        rl.close();
    }
}
