// channelkit/dist/cli/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.API_KEY_INFO = exports.COLORS = exports.LOGO = void 0;
exports.c = c;
exports.banner = banner;
exports.ask = ask;
exports.select = select;
exports.readEnvFile = readEnvFile;
exports.saveEnvVar = saveEnvVar;
exports.promptApiKey = promptApiKey;
const fs_1 = require("fs");
const path_1 = require("path");
exports.LOGO = `
  ╔═══════════════════════════════════════╗
  ║                                       ║
  ║     📱 ──┐                            ║
  ║     💬 ──┤──→ ChannelKit ──→ 🔗 App   ║
  ║     📞 ──┘                            ║
  ║                                       ║
  ║     Messaging ingress for your app    ║
  ║                                       ║
  ╚═══════════════════════════════════════╝
`;
exports.COLORS = {
    reset: '\x1b[0m',
    bright: '\x1b[1m',
    dim: '\x1b[2m',
    cyan: '\x1b[36m',
    green: '\x1b[32m',
    yellow: '\x1b[33m',
    magenta: '\x1b[35m',
    blue: '\x1b[34m',
    white: '\x1b[37m',
    gray: '\x1b[90m',
};
function c(color, text) {
    return `${exports.COLORS[color]}${text}${exports.COLORS.reset}`;
}
function banner() {
    console.log(c('cyan', exports.LOGO));
    console.log(c('dim', `  v0.1.0\n`));
}
async function ask(rl, question, defaultVal) {
    const suffix = defaultVal ? c('dim', ` (${defaultVal})`) : '';
    return new Promise((resolve) => {
        rl.question(`  ${c('green', '?')} ${question}${suffix} `, (answer) => {
            resolve(answer.trim() || defaultVal || '');
        });
    });
}
async function select(rl, question, options) {
    console.log(`\n  ${c('green', '?')} ${question}\n`);
    options.forEach((opt, i) => {
        console.log(`    ${c('cyan', `${i + 1})`)} ${opt}`);
    });
    console.log();
    while (true) {
        const answer = await ask(rl, `Choose ${c('dim', `[1-${options.length}]`)}:`);
        const num = parseInt(answer);
        if (num >= 1 && num <= options.length)
            return num - 1;
        console.log(c('yellow', '    Please enter a valid number'));
    }
}
// --- .env helpers ---
exports.API_KEY_INFO = {
    google: { envVar: 'GOOGLE_API_KEY', hint: 'Get one at https://console.cloud.google.com/apis/credentials' },
    whisper: { envVar: 'OPENAI_API_KEY', hint: 'Get one at https://platform.openai.com/api-keys' },
    deepgram: { envVar: 'DEEPGRAM_API_KEY', hint: 'Get one at https://console.deepgram.com' },
    elevenlabs: { envVar: 'ELEVENLABS_API_KEY', hint: 'Get one at https://elevenlabs.io/api' },
    openai: { envVar: 'OPENAI_API_KEY', hint: 'Get one at https://platform.openai.com/api-keys' },
};
function readEnvFile(envPath) {
    if (!(0, fs_1.existsSync)(envPath))
        return {};
    const content = (0, fs_1.readFileSync)(envPath, 'utf-8');
    const vars = {};
    for (const line of content.split('\n')) {
        const trimmed = line.trim();
        if (!trimmed || trimmed.startsWith('#'))
            continue;
        const eqIdx = trimmed.indexOf('=');
        if (eqIdx === -1)
            continue;
        const key = trimmed.slice(0, eqIdx).trim();
        let val = trimmed.slice(eqIdx + 1).trim();
        if ((val.startsWith('"') && val.endsWith('"')) || (val.startsWith("'") && val.endsWith("'"))) {
            val = val.slice(1, -1);
        }
        vars[key] = val;
    }
    return vars;
}
function saveEnvVar(envPath, key, value) {
    const existing = readEnvFile(envPath);
    if (existing[key])
        return;
    const line = `${key}=${value}\n`;
    (0, fs_1.appendFileSync)(envPath, line);
}
async function promptApiKey(rl, provider, configPath) {
    const info = exports.API_KEY_INFO[provider];
    if (!info)
        return;
    const envPath = (0, path_1.join)((0, path_1.dirname)((0, path_1.resolve)(configPath)), '.env');
    const envVars = readEnvFile(envPath);
    if (process.env[info.envVar] || envVars[info.envVar]) {
        console.log(c('dim', `\n  ✓ ${info.envVar} already set`));
        return;
    }
    console.log(c('dim', `\n  ${info.hint}`));
    const value = await ask(rl, `${info.envVar}:`);
    if (value) {
        saveEnvVar(envPath, info.envVar, value);
        process.env[info.envVar] = value;
        console.log(c('green', `  ✓ Saved to .env`));
        console.log(c('yellow', `  💡 Make sure .env is in your .gitignore`));
    }
}
