// channelkit/dist/channels/endpoint/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EndpointChannel = void 0;
const crypto_1 = require("crypto");
const base_1 = require("../base");
class EndpointChannel extends base_1.Channel {
    pendingResponses = new Map();
    constructor(name, config) {
        super(name, config);
    }
    get cfg() {
        return this.config;
    }
    async connect() {
        const method = (this.cfg.method || 'POST').toUpperCase();
        const mode = this.cfg.response_mode || 'sync';
        console.log(`[endpoint:${this.name}] Ready — ${method} /inbound/endpoint/${this.name} (${mode})`);
    }
    async disconnect() {
        for (const [key, pending] of this.pendingResponses) {
            clearTimeout(pending.timer);
            pending.resolve({ text: 'Channel disconnected' });
        }
        this.pendingResponses.clear();
    }
    async send(to, response) {
        const pending = this.pendingResponses.get(to);
        if (pending) {
            clearTimeout(pending.timer);
            this.pendingResponses.delete(to);
            pending.resolve(response);
        }
    }
    /**
     * Called by the inbound route handler.
     * Creates a UnifiedMessage, emits it, and optionally returns a promise
     * that resolves when the pipeline calls send() (sync mode).
     */
    handleRequest(body, query, headers, method) {
        const requestId = (0, crypto_1.randomUUID)();
        const fromId = `endpoint:${requestId}`;
        // Build text from request
        let text;
        if (method === 'GET') {
            text = query.text || JSON.stringify(query);
        }
        else {
            text = typeof body === 'string' ? body : (body != null ? JSON.stringify(body) : undefined);
        }
        const message = {
            id: requestId,
            channel: 'endpoint',
            from: fromId,
            type: 'text',
            text,
            timestamp: Math.floor(Date.now() / 1000),
            endpoint: {
                body: method !== 'GET' ? body : undefined,
                query,
                headers,
                method,
            },
        };
        let waitForResponse;
        if (this.cfg.response_mode === 'sync' || this.cfg.response_mode == null) {
            const timeout = (this.cfg.response_timeout || 30) * 1000;
            waitForResponse = new Promise((resolve) => {
                const timer = setTimeout(() => {
                    this.pendingResponses.delete(fromId);
                    resolve({ text: 'Response timeout', _error: true });
                }, timeout);
                this.pendingResponses.set(fromId, { resolve, timer });
            });
        }
        this.emitMessage(message);
        return { message, waitForResponse };
    }
}
exports.EndpointChannel = EndpointChannel;
