// channelkit/dist/core/groupStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GroupStore = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
const os_1 = require("os");
/** Default groups.json path: ~/.channelkit/data/groups.json (absolute, CWD-independent) */
const DEFAULT_GROUPS_PATH = (0, path_1.join)((0, os_1.homedir)(), '.channelkit', 'data', 'groups.json');
class GroupStore {
    filePath;
    groups = {};
    onChangeCallback;
    constructor(filePath = DEFAULT_GROUPS_PATH) {
        this.filePath = filePath;
        this.load();
    }
    load() {
        try {
            if ((0, fs_1.existsSync)(this.filePath)) {
                const raw = (0, fs_1.readFileSync)(this.filePath, 'utf-8');
                this.groups = JSON.parse(raw);
            }
        }
        catch {
            this.groups = {};
        }
    }
    save() {
        const dir = (0, path_1.dirname)(this.filePath);
        if (!(0, fs_1.existsSync)(dir))
            (0, fs_1.mkdirSync)(dir, { recursive: true });
        (0, fs_1.writeFileSync)(this.filePath, JSON.stringify(this.groups, null, 2));
        if (this.onChangeCallback)
            this.onChangeCallback({ ...this.groups });
    }
    /** Set a callback to be called whenever groups change (for remote sync). */
    onChange(callback) {
        this.onChangeCallback = callback;
    }
    /** Replace all groups with data fetched from a remote source. */
    replaceAll(groups) {
        this.groups = groups;
        const dir = (0, path_1.dirname)(this.filePath);
        if (!(0, fs_1.existsSync)(dir))
            (0, fs_1.mkdirSync)(dir, { recursive: true });
        (0, fs_1.writeFileSync)(this.filePath, JSON.stringify(this.groups, null, 2));
        // Don't trigger onChange to avoid sync loops
    }
    add(groupId, mapping) {
        this.groups[groupId] = mapping;
        this.save();
    }
    get(groupId) {
        return this.groups[groupId];
    }
    getAll() {
        return { ...this.groups };
    }
    findByUserAndService(userId, serviceName) {
        return Object.values(this.groups).find(g => g.userId === userId && g.serviceName === serviceName);
    }
}
exports.GroupStore = GroupStore;
