// channelkit/dist/core/remoteStore.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteStore = void 0;
/**
 * RemoteStore handles syncing all ChannelKit data with a remote REST endpoint.
 *
 * Remote API contract:
 *   GET  /config  → returns YAML config as text/plain
 *   PUT  /config  → accepts YAML config as text/plain body
 *   GET  /auth    → returns auth directory as application/zip
 *   PUT  /auth    → accepts auth directory as application/zip body
 *   GET  /groups  → returns group mappings as application/json
 *   PUT  /groups  → accepts group mappings as application/json body
 */
class RemoteStore {
    config;
    debounceTimer = null;
    syncIntervalMs;
    endpoint;
    authHeader;
    constructor(config) {
        this.config = config;
        this.endpoint = (config.endpoint || '').replace(/\/$/, '');
        this.authHeader = config.auth_header;
        this.syncIntervalMs = (config.sync_interval || 30) * 1000;
    }
    headers(extra) {
        const h = {};
        if (this.authHeader)
            h['Authorization'] = this.authHeader;
        if (extra)
            Object.assign(h, extra);
        return h;
    }
    // ── Config ──────────────────────────────────────────────────────────
    /**
     * Fetch config YAML from the remote endpoint. Returns null if unreachable or empty.
     */
    async fetchConfig() {
        if (!this.endpoint)
            return null;
        try {
            const res = await fetch(`${this.endpoint}/config`, {
                headers: this.headers({ 'Accept': 'text/plain' }),
            });
            if (res.status === 404)
                return null; // no config stored yet
            if (!res.ok) {
                console.warn(`[remote-store] Failed to fetch config: HTTP ${res.status}`);
                return null;
            }
            const text = await res.text();
            return text || null;
        }
        catch (err) {
            console.warn(`[remote-store] Failed to fetch config: ${err.message}`);
            return null;
        }
    }
    /**
     * Push config YAML to the remote endpoint.
     */
    async pushConfig(yaml) {
        if (!this.endpoint)
            return false;
        try {
            const res = await fetch(`${this.endpoint}/config`, {
                method: 'PUT',
                headers: this.headers({ 'Content-Type': 'text/plain' }),
                body: yaml,
            });
            if (!res.ok) {
                console.warn(`[remote-store] Failed to push config: HTTP ${res.status}`);
                return false;
            }
            console.log('[remote-store] Config synced to remote');
            return true;
        }
        catch (err) {
            console.warn(`[remote-store] Failed to push config: ${err.message}`);
            return false;
        }
    }
    // ── Auth ─────────────────────────────────────────────────────────────
    /**
     * Fetch auth data as a ZIP buffer from the remote endpoint.
     */
    async fetchAuth() {
        if (!this.endpoint)
            return null;
        try {
            const res = await fetch(`${this.endpoint}/auth`, {
                headers: this.headers({ 'Accept': 'application/zip' }),
            });
            if (res.status === 404)
                return null; // no auth stored yet
            if (!res.ok) {
                console.warn(`[remote-store] Failed to fetch auth: HTTP ${res.status}`);
                return null;
            }
            const arrayBuf = await res.arrayBuffer();
            const buf = Buffer.from(arrayBuf);
            return buf.length > 0 ? buf : null;
        }
        catch (err) {
            console.warn(`[remote-store] Failed to fetch auth: ${err.message}`);
            return null;
        }
    }
    /**
     * Push auth data as a ZIP buffer to the remote endpoint.
     */
    async pushAuth(zipBuffer) {
        if (!this.endpoint)
            return false;
        try {
            const res = await fetch(`${this.endpoint}/auth`, {
                method: 'PUT',
                headers: this.headers({ 'Content-Type': 'application/zip' }),
                body: zipBuffer,
            });
            if (!res.ok) {
                console.warn(`[remote-store] Failed to push auth: HTTP ${res.status}`);
                return false;
            }
            console.log('[remote-store] Auth synced to remote');
            return true;
        }
        catch (err) {
            console.warn(`[remote-store] Failed to push auth: ${err.message}`);
            return false;
        }
    }
    // ── Groups ──────────────────────────────────────────────────────────
    /**
     * Fetch groups from the remote endpoint. Returns null if unreachable.
     */
    async fetchGroups() {
        if (!this.endpoint)
            return null;
        try {
            const res = await fetch(`${this.endpoint}/groups`, {
                headers: this.headers({ 'Accept': 'application/json' }),
            });
            if (res.status === 404)
                return null;
            if (!res.ok) {
                console.warn(`[remote-store] Failed to fetch groups: HTTP ${res.status}`);
                return null;
            }
            const data = await res.json();
            return data;
        }
        catch (err) {
            console.warn(`[remote-store] Failed to fetch groups: ${err.message}`);
            return null;
        }
    }
    /**
     * Push groups to the remote endpoint (debounced).
     */
    pushGroups(groups) {
        if (!this.endpoint)
            return;
        if (this.debounceTimer)
            clearTimeout(this.debounceTimer);
        this.debounceTimer = setTimeout(() => this.doPushGroups(groups), this.syncIntervalMs);
    }
    /**
     * Immediately push groups to the remote endpoint.
     */
    async pushGroupsNow(groups) {
        return this.doPushGroups(groups);
    }
    async doPushGroups(groups) {
        if (!this.endpoint)
            return false;
        try {
            const res = await fetch(`${this.endpoint}/groups`, {
                method: 'PUT',
                headers: this.headers({ 'Content-Type': 'application/json' }),
                body: JSON.stringify(groups),
            });
            if (!res.ok) {
                console.warn(`[remote-store] Failed to push groups: HTTP ${res.status}`);
                return false;
            }
            console.log('[remote-store] Groups synced to remote');
            return true;
        }
        catch (err) {
            console.warn(`[remote-store] Failed to push groups: ${err.message}`);
            return false;
        }
    }
    // ── Helpers ─────────────────────────────────────────────────────────
    /**
     * Zip a directory into a buffer using the system `zip` command.
     */
    static async zipDirectory(dirPath) {
        const { existsSync } = await Promise.resolve().then(() => __importStar(require('fs')));
        const { readFile, rm } = await Promise.resolve().then(() => __importStar(require('fs/promises')));
        const { execFile } = await Promise.resolve().then(() => __importStar(require('child_process')));
        const { tmpdir } = await Promise.resolve().then(() => __importStar(require('os')));
        const { join } = await Promise.resolve().then(() => __importStar(require('path')));
        const { randomBytes } = await Promise.resolve().then(() => __importStar(require('crypto')));
        if (!existsSync(dirPath))
            return null;
        // Check if directory has any files
        const { readdirSync } = await Promise.resolve().then(() => __importStar(require('fs')));
        const entries = readdirSync(dirPath);
        if (entries.length === 0)
            return null;
        const tmpPath = join(tmpdir(), `ck-auth-${randomBytes(4).toString('hex')}.zip`);
        try {
            await new Promise((resolve, reject) => {
                execFile('zip', ['-r', '-q', tmpPath, '.'], { cwd: dirPath }, (err, _stdout, stderr) => {
                    if (err)
                        reject(new Error(stderr || err.message));
                    else
                        resolve();
                });
            });
            const buf = await readFile(tmpPath);
            await rm(tmpPath, { force: true });
            return buf;
        }
        catch (err) {
            await rm(tmpPath, { force: true }).catch(() => { });
            console.warn(`[remote-store] Failed to zip auth dir: ${err.message}`);
            return null;
        }
    }
    /**
     * Unzip a buffer into a directory using the system `unzip` command.
     */
    static async unzipToDirectory(zipBuffer, dirPath) {
        const { mkdirSync, existsSync } = await Promise.resolve().then(() => __importStar(require('fs')));
        const { writeFile, rm } = await Promise.resolve().then(() => __importStar(require('fs/promises')));
        const { execFile } = await Promise.resolve().then(() => __importStar(require('child_process')));
        const { tmpdir } = await Promise.resolve().then(() => __importStar(require('os')));
        const { join } = await Promise.resolve().then(() => __importStar(require('path')));
        const { randomBytes } = await Promise.resolve().then(() => __importStar(require('crypto')));
        if (!existsSync(dirPath))
            mkdirSync(dirPath, { recursive: true });
        const tmpPath = join(tmpdir(), `ck-auth-${randomBytes(4).toString('hex')}.zip`);
        try {
            await writeFile(tmpPath, zipBuffer);
            await new Promise((resolve, reject) => {
                execFile('unzip', ['-o', '-q', tmpPath, '-d', dirPath], (err, _stdout, stderr) => {
                    if (err)
                        reject(new Error(stderr || err.message));
                    else
                        resolve();
                });
            });
            await rm(tmpPath, { force: true });
            return true;
        }
        catch (err) {
            await rm(tmpPath, { force: true }).catch(() => { });
            console.warn(`[remote-store] Failed to unzip auth: ${err.message}`);
            return false;
        }
    }
}
exports.RemoteStore = RemoteStore;
