// channelkit/dist/core/tunnel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TunnelManager = void 0;
const cloudflared_1 = require("cloudflared");
class TunnelManager {
    config;
    targetPort;
    publicUrl = null;
    tunnel = null;
    constructor(config, targetPort) {
        this.config = config;
        this.targetPort = targetPort;
    }
    getPublicUrl() {
        return this.publicUrl;
    }
    async start() {
        if (this.config.public_url && !this.config.token) {
            // Manual public URL without a token — just use it directly
            this.publicUrl = this.config.public_url.replace(/\/$/, '');
            console.log(`📡 Public URL: ${this.publicUrl}`);
            return;
        }
        if (this.config.provider === 'cloudflared' || !this.config.provider) {
            if (this.config.token) {
                await this.startWithToken();
            }
            else {
                await this.startQuickTunnel();
            }
            return;
        }
    }
    async stop() {
        if (this.tunnel) {
            this.tunnel.stop();
            this.tunnel = null;
        }
        this.publicUrl = null;
    }
    /** Named tunnel with a Cloudflare Tunnel token — gives a stable URL. */
    async startWithToken() {
        // The public URL is pre-configured in the Cloudflare dashboard
        if (this.config.public_url) {
            this.publicUrl = this.config.public_url.replace(/\/$/, '');
        }
        return new Promise((resolve, reject) => {
            const t = cloudflared_1.Tunnel.withToken(this.config.token);
            this.tunnel = t;
            const timeout = setTimeout(() => {
                // If no error after 15s, assume the tunnel is running
                console.log(`📡 Tunnel connected — Public URL: ${this.publicUrl || '(check Cloudflare dashboard)'}`);
                resolve();
            }, 15000);
            t.on('connected', () => {
                clearTimeout(timeout);
                console.log(`📡 Tunnel connected — Public URL: ${this.publicUrl || '(check Cloudflare dashboard)'}`);
                resolve();
            });
            t.on('error', (err) => {
                clearTimeout(timeout);
                reject(new Error(`Failed to start cloudflared: ${err.message}`));
            });
            t.on('exit', (code) => {
                clearTimeout(timeout);
                reject(new Error(`cloudflared exited with code ${code} before connecting`));
            });
        });
    }
    /** Quick tunnel (trycloudflare.com) — random URL each time. */
    async startQuickTunnel() {
        const targetUrl = `http://localhost:${this.targetPort}`;
        return new Promise((resolve, reject) => {
            const t = cloudflared_1.Tunnel.quick(targetUrl);
            this.tunnel = t;
            const timeout = setTimeout(() => {
                reject(new Error('Timed out waiting for cloudflared tunnel URL (15s)'));
            }, 15000);
            t.on('url', (url) => {
                this.publicUrl = url;
                clearTimeout(timeout);
                console.log(`📡 Public URL: ${this.publicUrl}`);
                resolve();
            });
            t.on('error', (err) => {
                clearTimeout(timeout);
                reject(new Error(`Failed to start cloudflared: ${err.message}`));
            });
            t.on('exit', (code) => {
                if (!this.publicUrl) {
                    clearTimeout(timeout);
                    reject(new Error(`cloudflared exited with code ${code} before providing a URL`));
                }
            });
        });
    }
}
exports.TunnelManager = TunnelManager;
