// channelkit/dist/core/updater.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Updater = void 0;
const child_process_1 = require("child_process");
const fs_1 = require("fs");
const path_1 = require("path");
const restart_1 = require("./restart");
const PKG_NAME = 'channelkit';
class Updater {
    channels;
    onLog;
    lastStatus = null;
    autoUpdateTimer = null;
    projectRoot;
    updating = false;
    mode;
    constructor(channels, onLog) {
        this.channels = channels;
        this.onLog = onLog;
        this.projectRoot = process.cwd();
        this.mode = this.detectMode();
        this.log(`Running in ${this.mode} mode`);
    }
    detectMode() {
        // If there's a .git directory in the project root, it's a git clone
        if ((0, fs_1.existsSync)((0, path_1.join)(this.projectRoot, '.git')))
            return 'git';
        // Detect npx: running from a _npx cache directory
        if (__dirname.includes('_npx') || __dirname.includes('.npm/_npx'))
            return 'npx';
        return 'npm';
    }
    getMode() {
        return this.mode;
    }
    log(msg) {
        const formatted = `[updater] ${msg}`;
        console.log(formatted);
        this.onLog?.(formatted);
    }
    exec(cmd, timeout = 30000) {
        return (0, child_process_1.execSync)(cmd, { cwd: this.projectRoot, encoding: 'utf-8', timeout }).trim();
    }
    getCurrentVersion() {
        if (this.mode === 'git') {
            try {
                return this.exec('git rev-parse --short HEAD');
            }
            catch {
                return 'unknown';
            }
        }
        // npm mode — read version from our own package.json
        try {
            const pkgPath = (0, path_1.join)(__dirname, '..', '..', 'package.json');
            return require(pkgPath).version || 'unknown';
        }
        catch {
            return 'unknown';
        }
    }
    getLastStatus() {
        return this.lastStatus;
    }
    async checkForUpdate() {
        if (this.mode === 'git') {
            return this.checkGitUpdate();
        }
        return this.checkNpmUpdate();
    }
    async checkGitUpdate() {
        const currentVersion = this.getCurrentVersion();
        try {
            this.exec('git fetch origin main', 30000);
        }
        catch (err) {
            this.log(`Failed to fetch from origin: ${err.message}`);
            this.lastStatus = {
                mode: 'git',
                currentVersion,
                latestVersion: currentVersion,
                updateAvailable: false,
                behindCount: 0,
                lastChecked: Date.now(),
            };
            return this.lastStatus;
        }
        const latestVersion = this.exec('git rev-parse --short origin/main');
        const currentFull = this.exec('git rev-parse HEAD');
        const remoteFull = this.exec('git rev-parse origin/main');
        let behindCount = 0;
        try {
            behindCount = parseInt(this.exec('git rev-list HEAD..origin/main --count')) || 0;
        }
        catch { }
        const updateAvailable = currentFull !== remoteFull;
        this.lastStatus = {
            mode: 'git',
            currentVersion,
            latestVersion,
            updateAvailable,
            behindCount,
            lastChecked: Date.now(),
        };
        if (updateAvailable) {
            this.log(`Update available: ${currentVersion} -> ${latestVersion} (${behindCount} commit(s) behind)`);
        }
        return this.lastStatus;
    }
    async checkNpmUpdate() {
        const currentVersion = this.getCurrentVersion();
        let latestVersion = currentVersion;
        try {
            latestVersion = this.exec(`npm view ${PKG_NAME} version`, 15000);
        }
        catch (err) {
            this.log(`Failed to check npm registry: ${err.message}`);
            this.lastStatus = {
                mode: 'npm',
                currentVersion,
                latestVersion: currentVersion,
                updateAvailable: false,
                behindCount: 0,
                lastChecked: Date.now(),
            };
            return this.lastStatus;
        }
        const updateAvailable = currentVersion !== latestVersion;
        this.lastStatus = {
            mode: 'npm',
            currentVersion,
            latestVersion,
            updateAvailable,
            behindCount: updateAvailable ? 1 : 0,
            lastChecked: Date.now(),
        };
        if (updateAvailable) {
            this.log(`Update available: ${currentVersion} -> ${latestVersion}`);
        }
        return this.lastStatus;
    }
    async performUpdate() {
        if (this.updating) {
            return { success: false, previousVersion: '', newVersion: '', error: 'Update already in progress' };
        }
        if (this.mode === 'npx') {
            const currentVersion = this.getCurrentVersion();
            return {
                success: false,
                previousVersion: currentVersion,
                newVersion: currentVersion,
                error: `Auto-update is not supported when running via npx. Run: npx channelkit@latest`,
            };
        }
        this.updating = true;
        if (this.mode === 'git') {
            return this.performGitUpdate();
        }
        return this.performNpmUpdate();
    }
    async performGitUpdate() {
        const previousVersion = this.getCurrentVersion();
        try {
            const dirty = this.exec('git status --porcelain');
            if (dirty) {
                this.updating = false;
                return { success: false, previousVersion, newVersion: previousVersion, error: 'Working tree has uncommitted changes. Commit or stash them first.' };
            }
            this.log('Pulling latest changes from origin/main...');
            this.exec('git pull origin main', 60000);
            this.log('Installing dependencies...');
            this.exec('npm install', 120000);
            this.log('Building...');
            this.exec('npm run build', 120000);
            const newVersion = this.getCurrentVersion();
            this.log(`Update complete: ${previousVersion} -> ${newVersion}. Restarting...`);
            setTimeout(() => (0, restart_1.restartProcess)(this.channels), 500);
            return { success: true, previousVersion, newVersion };
        }
        catch (err) {
            this.updating = false;
            this.log(`Update failed: ${err.message}`);
            return { success: false, previousVersion, newVersion: previousVersion, error: err.message };
        }
    }
    async performNpmUpdate() {
        const previousVersion = this.getCurrentVersion();
        try {
            this.log(`Updating ${PKG_NAME} via npm...`);
            // Detect if installed globally or locally
            const isGlobal = this.isGlobalInstall();
            const installCmd = isGlobal
                ? `npm install -g ${PKG_NAME}@latest`
                : `npm install ${PKG_NAME}@latest`;
            this.log(`Running: ${installCmd}`);
            this.exec(installCmd, 120000);
            const newVersion = this.exec(`npm view ${PKG_NAME} version`, 15000);
            this.log(`Update complete: ${previousVersion} -> ${newVersion}. Restarting...`);
            setTimeout(() => (0, restart_1.restartProcess)(this.channels), 500);
            return { success: true, previousVersion, newVersion };
        }
        catch (err) {
            this.updating = false;
            this.log(`Update failed: ${err.message}`);
            return { success: false, previousVersion, newVersion: previousVersion, error: err.message };
        }
    }
    isGlobalInstall() {
        try {
            const globalPrefix = this.exec('npm prefix -g');
            // If our package's path starts with the global prefix, it's a global install
            return __dirname.startsWith(globalPrefix);
        }
        catch {
            return false;
        }
    }
    /** Check for updates periodically without installing (notify only). */
    startUpdateCheck(intervalMinutes) {
        if (this.autoUpdateTimer) {
            clearInterval(this.autoUpdateTimer);
        }
        const intervalMs = intervalMinutes * 60 * 1000;
        this.log(`Update check enabled (notify only): checking every ${intervalMinutes} minute(s)`);
        // Initial check after 10s
        setTimeout(() => this.notifyUpdateCheck(), 10000);
        this.autoUpdateTimer = setInterval(() => this.notifyUpdateCheck(), intervalMs);
    }
    async notifyUpdateCheck() {
        try {
            const status = await this.checkForUpdate();
            if (status.updateAvailable) {
                let updateCmd;
                if (status.mode === 'npx') {
                    updateCmd = `npx ${PKG_NAME}@latest`;
                }
                else if (status.mode === 'npm') {
                    updateCmd = `npm update -g ${PKG_NAME}`;
                }
                else {
                    updateCmd = 'git pull && npm run build';
                }
                this.log(`Update available: ${status.currentVersion} → ${status.latestVersion}. Run: ${updateCmd}`);
                console.log(`\n  ╭─────────────────────────────────────────────╮`);
                console.log(`  │  🆕 New version available!                   │`);
                console.log(`  │  ${status.currentVersion} → ${status.latestVersion}${' '.repeat(Math.max(0, 31 - status.currentVersion.length - status.latestVersion.length))}│`);
                console.log(`  │  Update: ${updateCmd}${' '.repeat(Math.max(0, 35 - updateCmd.length))}│`);
                console.log(`  ╰─────────────────────────────────────────────╯\n`);
            }
        }
        catch (err) {
            this.log(`Update check failed: ${err.message}`);
        }
    }
    startAutoUpdate(intervalMinutes) {
        if (this.autoUpdateTimer) {
            clearInterval(this.autoUpdateTimer);
        }
        const intervalMs = intervalMinutes * 60 * 1000;
        this.log(`Auto-update enabled: checking every ${intervalMinutes} minute(s)`);
        setTimeout(() => this.autoUpdateCheck(), 10000);
        this.autoUpdateTimer = setInterval(() => this.autoUpdateCheck(), intervalMs);
    }
    stopAutoUpdate() {
        if (this.autoUpdateTimer) {
            clearInterval(this.autoUpdateTimer);
            this.autoUpdateTimer = null;
            this.log('Auto-update disabled');
        }
    }
    async autoUpdateCheck() {
        try {
            const status = await this.checkForUpdate();
            if (status.updateAvailable) {
                this.log(`Auto-update: new version detected (${status.latestVersion}). Updating...`);
                await this.performUpdate();
            }
        }
        catch (err) {
            this.log(`Auto-update check failed: ${err.message}`);
        }
    }
}
exports.Updater = Updater;
