// channelkit/dist/onboarding/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Onboarding = void 0;
const groupStore_1 = require("../core/groupStore");
class Onboarding {
    config;
    whatsappChannel;
    telegramChannel;
    groupStore;
    // Track telegram user→service mappings (chatId → serviceName)
    telegramMappings = new Map();
    constructor(config, whatsappChannel, telegramChannel) {
        this.config = config;
        this.whatsappChannel = whatsappChannel;
        this.telegramChannel = telegramChannel;
        this.groupStore = new groupStore_1.GroupStore();
    }
    getGroupStore() {
        return this.groupStore;
    }
    /**
     * Handle a direct message for onboarding. Returns true if handled.
     * @param unmatchedPolicy - what to do when no code matches ('list' = send menu, 'ignore' or undefined = silent)
     */
    async handleDirectMessage(message, unmatchedPolicy) {
        if (message.groupId)
            return false;
        if (message.channel === 'whatsapp') {
            return this.handleWhatsApp(message, unmatchedPolicy);
        }
        if (message.channel === 'telegram') {
            return this.handleTelegram(message, unmatchedPolicy);
        }
        return false;
    }
    /**
     * Check if a Telegram chat has a service mapping. Returns the webhook URL if so.
     */
    getTelegramServiceWebhook(chatId) {
        const serviceName = this.telegramMappings.get(chatId);
        if (!serviceName)
            return undefined;
        const code = (this.config.codes || []).find(c => c.name === serviceName);
        return code?.webhook;
    }
    // --- WhatsApp ---
    async handleWhatsApp(message, unmatchedPolicy) {
        if (!this.whatsappChannel)
            return false;
        const codes = this.getCodesForChannel(message.channelName || message.channel);
        const text = (message.text || '').trim().toUpperCase();
        const matched = codes.find(c => c.code.toUpperCase() === text);
        if (matched) {
            const existing = this.groupStore.findByUserAndService(message.from, matched.name);
            if (existing) {
                await this.whatsappChannel.sendToJid(message.from, `You're already connected to ${matched.name}! Check the group "${matched.name}" in your chats.`);
                return true;
            }
            const now = Math.floor(Date.now() / 1000);
            if (message.timestamp && now - message.timestamp > 30) {
                console.log(`[onboarding] Skipping old message from ${message.from} (${now - message.timestamp}s old)`);
                return true;
            }
            await this.createWhatsAppGroup(message.from, matched, message.senderName);
            return true;
        }
        // No match — send menu only if the channel's unmatched policy is 'list'
        if (unmatchedPolicy === 'list' && codes.length > 0) {
            const codeList = codes.map(c => c.code).join(' or ');
            await this.whatsappChannel.sendToJid(message.from, `Available services: send ${codeList} to connect`);
        }
        return true;
    }
    async createWhatsAppGroup(userJid, service, senderName) {
        if (!this.whatsappChannel)
            return;
        const phone = userJid.replace(/@s\.whatsapp\.net$/, '');
        const groupName = `${service.name} - ${senderName || phone}`;
        try {
            console.log(`[onboarding] Creating group "${groupName}" with participant: ${userJid}`);
            const group = await this.whatsappChannel.createGroup(groupName, [userJid]);
            this.groupStore.add(group.id, {
                groupId: group.id,
                serviceName: service.name,
                webhook: service.webhook,
                userId: userJid,
                createdAt: Date.now(),
            });
            await this.whatsappChannel.sendToJid(group.id, `Welcome to ${service.name}! 🎉\nAll messages in this group will be forwarded to the service.`);
            console.log(`[onboarding] Created group "${groupName}" (${group.id}) for service ${service.name}`);
        }
        catch (err) {
            console.error(`[onboarding] Failed to create group:`, err);
            await this.whatsappChannel.sendToJid(userJid, `Sorry, failed to set up ${service.name}. Please try again.`);
        }
    }
    // --- Telegram ---
    async handleTelegram(message, unmatchedPolicy) {
        if (!this.telegramChannel)
            return false;
        const codes = this.getCodesForChannel(message.channelName || message.channel);
        if (codes.length === 0)
            return false;
        // Support both plain code and /start CODE
        let text = (message.text || '').trim().toUpperCase();
        if (text.startsWith('/START ')) {
            text = text.slice(7).trim();
        }
        const matched = codes.find(c => c.code.toUpperCase() === text);
        if (matched) {
            // Check if already connected
            const existingService = this.telegramMappings.get(message.from);
            if (existingService === matched.name) {
                await this.telegramChannel.sendToChat(message.from, `You're already connected to ${matched.name}! Just send messages here.`);
                return true;
            }
            // Map this chat to the service
            this.telegramMappings.set(message.from, matched.name);
            await this.telegramChannel.sendToChat(message.from, `Welcome to ${matched.name}! 🎉\nAll messages here will be forwarded to the service.`);
            console.log(`[onboarding] Telegram user ${message.senderName || message.from} connected to ${matched.name}`);
            return true;
        }
        // If already mapped to a service, don't show menu — let it route
        if (this.telegramMappings.has(message.from)) {
            return false;
        }
        // No match, no mapping — send menu only if the channel's unmatched policy is 'list'
        if (unmatchedPolicy === 'list' && codes.length > 0) {
            const codeList = codes.map(c => `\`${c.code}\` → ${c.name}`).join('\n');
            await this.telegramChannel.sendToChat(message.from, `Available services:\n${codeList}\n\nSend a code to connect.`);
        }
        return true;
    }
    // --- Helpers ---
    getCodesForChannel(channel) {
        return (this.config.codes || []).filter(c => {
            if (!c.channels || c.channels.length === 0)
                return true; // all channels
            return c.channels.includes(channel);
        });
    }
}
exports.Onboarding = Onboarding;
