// channelkit/dist/channels/email/gmail.js
"use strict";
/**
 * Gmail Channel — OAuth2 + polling
 *
 * Flow:
 * 1. First run: opens browser for OAuth consent → saves refresh token
 * 2. Polls for new messages at configured interval
 * 3. Sends replies via Gmail API
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GmailChannel = void 0;
const base_1 = require("../base");
const fs_1 = require("fs");
const path_1 = require("path");
const paths_1 = require("../../paths");
class GmailChannel extends base_1.Channel {
    tokens = null;
    tokenPath;
    pollTimer = null;
    lastHistoryId = null;
    processedIds = new Set();
    constructor(name, config) {
        super(name, config);
        this.tokenPath = (0, path_1.join)(paths_1.DEFAULT_AUTH_DIR, `gmail-${name}.json`);
    }
    get cfg() {
        return this.config;
    }
    async connect() {
        // Load saved tokens
        if ((0, fs_1.existsSync)(this.tokenPath)) {
            this.tokens = JSON.parse((0, fs_1.readFileSync)(this.tokenPath, 'utf-8'));
        }
        if (!this.tokens?.refresh_token) {
            // Need OAuth flow
            await this.startOAuthFlow();
        }
        else {
            // Refresh access token
            await this.refreshAccessToken();
        }
        // Get initial history ID
        const profile = await this.gmailApi('GET', '/users/me/profile');
        this.lastHistoryId = profile.historyId;
        // Start polling
        const interval = (this.cfg.poll_interval || 30) * 1000;
        console.log(`✅ Gmail connected: ${this.name} (polling every ${interval / 1000}s)`);
        // Poll immediately once, then on interval
        this.poll();
        this.pollTimer = setInterval(() => this.poll(), interval);
    }
    async disconnect() {
        if (this.pollTimer) {
            clearInterval(this.pollTimer);
            this.pollTimer = null;
        }
    }
    async send(to, response) {
        if (!this.tokens)
            return;
        await this.ensureValidToken();
        const subject = response.email?.subject || '';
        const body = response.email?.html || response.text || '';
        const isHtml = !!response.email?.html;
        // Build RFC 2822 message
        const headers = [
            `To: ${to}`,
            `Subject: ${subject}`,
            `MIME-Version: 1.0`,
            `Content-Type: ${isHtml ? 'text/html' : 'text/plain'}; charset=utf-8`,
        ].join('\r\n');
        const raw = Buffer.from(`${headers}\r\n\r\n${body}`).toString('base64url');
        await this.gmailApi('POST', '/users/me/messages/send', { raw });
    }
    /**
     * Send a reply to a specific thread
     */
    async sendReply(to, response, threadId, inReplyTo) {
        if (!this.tokens)
            return;
        await this.ensureValidToken();
        const subject = response.email?.subject || '';
        const body = response.email?.html || response.text || '';
        const isHtml = !!response.email?.html;
        const headers = [
            `To: ${to}`,
            `Subject: ${subject}`,
            `MIME-Version: 1.0`,
            `Content-Type: ${isHtml ? 'text/html' : 'text/plain'}; charset=utf-8`,
            ...(inReplyTo ? [`In-Reply-To: ${inReplyTo}`, `References: ${inReplyTo}`] : []),
        ].join('\r\n');
        const raw = Buffer.from(`${headers}\r\n\r\n${body}`).toString('base64url');
        await this.gmailApi('POST', '/users/me/messages/send', { raw, threadId });
    }
    // --- OAuth Flow ---
    async startOAuthFlow() {
        const http = await Promise.resolve().then(() => __importStar(require('http')));
        const { exec: execCmd } = await Promise.resolve().then(() => __importStar(require('child_process')));
        // Start a temporary HTTP server on a random port to receive the OAuth callback
        const { code, redirectUri } = await new Promise((resolve, reject) => {
            const server = http.createServer((req, res) => {
                const url = new URL(req.url || '/', `http://localhost`);
                const authCode = url.searchParams.get('code');
                const error = url.searchParams.get('error');
                if (error) {
                    res.writeHead(200, { 'Content-Type': 'text/html' });
                    res.end('<html><body><h2>❌ Authorization failed</h2><p>You can close this tab.</p></body></html>');
                    server.close();
                    reject(new Error(`OAuth error: ${error}`));
                    return;
                }
                if (authCode) {
                    res.writeHead(200, { 'Content-Type': 'text/html' });
                    res.end('<html><body><h2>✅ Authorization successful!</h2><p>You can close this tab and return to the terminal.</p></body></html>');
                    server.close();
                    resolve({ code: authCode, redirectUri: localRedirectUri });
                }
                else {
                    res.writeHead(400, { 'Content-Type': 'text/plain' });
                    res.end('Missing code parameter');
                }
            });
            let localRedirectUri = '';
            server.listen(0, '127.0.0.1', () => {
                const addr = server.address();
                localRedirectUri = `http://localhost:${addr.port}`;
                const authUrl = `https://accounts.google.com/o/oauth2/v2/auth?` +
                    `client_id=${encodeURIComponent(this.cfg.client_id)}` +
                    `&redirect_uri=${encodeURIComponent(localRedirectUri)}` +
                    `&response_type=code` +
                    `&scope=${encodeURIComponent('https://www.googleapis.com/auth/gmail.modify')}` +
                    `&access_type=offline` +
                    `&prompt=consent`;
                console.log(`\n  📧 Gmail OAuth required for channel "${this.name}"`);
                console.log(`\n  Opening browser for authorization...\n`);
                // Try to open browser automatically
                execCmd(`open "${authUrl}" 2>/dev/null || xdg-open "${authUrl}" 2>/dev/null`, (err) => {
                    if (err) {
                        console.log(`  Could not open browser automatically. Open this URL:\n`);
                        console.log(`  ${authUrl}\n`);
                    }
                });
                // Timeout after 2 minutes
                setTimeout(() => {
                    server.close();
                    reject(new Error('OAuth timeout — no callback received within 2 minutes'));
                }, 120000);
            });
        });
        // Exchange code for tokens
        const res = await fetch('https://oauth2.googleapis.com/token', {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            body: new URLSearchParams({
                code,
                client_id: this.cfg.client_id,
                client_secret: this.cfg.client_secret,
                redirect_uri: redirectUri,
                grant_type: 'authorization_code',
            }),
        });
        if (!res.ok) {
            const err = await res.text();
            throw new Error(`Gmail OAuth failed: ${err}`);
        }
        const data = await res.json();
        this.tokens = {
            access_token: data.access_token,
            refresh_token: data.refresh_token,
            expiry: Date.now() + (data.expires_in * 1000),
        };
        this.saveTokens();
        console.log(`  ✅ Gmail authenticated successfully!\n`);
    }
    async refreshAccessToken() {
        if (!this.tokens?.refresh_token)
            throw new Error('No refresh token');
        const res = await fetch('https://oauth2.googleapis.com/token', {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            body: new URLSearchParams({
                refresh_token: this.tokens.refresh_token,
                client_id: this.cfg.client_id,
                client_secret: this.cfg.client_secret,
                grant_type: 'refresh_token',
            }),
        });
        if (!res.ok) {
            const err = await res.text();
            throw new Error(`Gmail token refresh failed: ${err}`);
        }
        const data = await res.json();
        this.tokens.access_token = data.access_token;
        this.tokens.expiry = Date.now() + (data.expires_in * 1000);
        this.saveTokens();
    }
    async ensureValidToken() {
        if (!this.tokens)
            throw new Error('Not authenticated');
        if (Date.now() > this.tokens.expiry - 60000) {
            await this.refreshAccessToken();
        }
    }
    saveTokens() {
        const dir = (0, path_1.dirname)(this.tokenPath);
        if (!(0, fs_1.existsSync)(dir))
            (0, fs_1.mkdirSync)(dir, { recursive: true });
        (0, fs_1.writeFileSync)(this.tokenPath, JSON.stringify(this.tokens, null, 2));
    }
    // --- Gmail API ---
    async gmailApi(method, path, body) {
        await this.ensureValidToken();
        const res = await fetch(`https://gmail.googleapis.com/gmail/v1${path}`, {
            method,
            headers: {
                'Authorization': `Bearer ${this.tokens.access_token}`,
                ...(body ? { 'Content-Type': 'application/json' } : {}),
            },
            ...(body ? { body: JSON.stringify(body) } : {}),
        });
        if (!res.ok) {
            const err = await res.text();
            throw new Error(`Gmail API ${method} ${path}: ${res.status} ${err}`);
        }
        return res.json();
    }
    // --- Polling ---
    async poll() {
        try {
            await this.ensureValidToken();
            // List new messages
            const labels = this.cfg.labels || ['INBOX'];
            const labelQuery = labels.map(l => `label:${l}`).join(' OR ');
            const query = `is:unread ${labelQuery}`;
            const list = await this.gmailApi('GET', `/users/me/messages?q=${encodeURIComponent(query)}&maxResults=10`);
            if (!list.messages || list.messages.length === 0)
                return;
            for (const { id } of list.messages) {
                if (this.processedIds.has(id))
                    continue;
                this.processedIds.add(id);
                // Keep processed set bounded
                if (this.processedIds.size > 1000) {
                    const arr = [...this.processedIds];
                    this.processedIds = new Set(arr.slice(-500));
                }
                const msg = await this.gmailApi('GET', `/users/me/messages/${id}?format=full`);
                const unified = this.toUnified(msg);
                if (unified) {
                    this.emitMessage(unified);
                    // Mark as read
                    await this.gmailApi('POST', `/users/me/messages/${id}/modify`, {
                        removeLabelIds: ['UNREAD'],
                    });
                }
            }
        }
        catch (err) {
            console.error(`[gmail:${this.name}] Poll error:`, err);
        }
    }
    toUnified(msg) {
        const headers = msg.payload?.headers || [];
        const getHeader = (name) => headers.find((h) => h.name.toLowerCase() === name.toLowerCase())?.value;
        const from = getHeader('From') || '';
        const subject = getHeader('Subject') || '';
        const to = getHeader('To') || '';
        const cc = getHeader('Cc')?.split(',').map((s) => s.trim()) || [];
        const messageId = getHeader('Message-ID') || '';
        // Extract body
        let textBody = '';
        let htmlBody = '';
        this.extractBody(msg.payload, (text, html) => {
            if (text)
                textBody = text;
            if (html)
                htmlBody = html;
        });
        // Extract sender email
        const emailMatch = from.match(/<(.+?)>/) || [null, from];
        const senderEmail = emailMatch[1] || from;
        const senderName = from.replace(/<.+?>/, '').trim().replace(/^"|"$/g, '');
        return {
            id: msg.id,
            channel: 'email',
            from: senderEmail,
            type: 'email',
            text: textBody || subject,
            timestamp: Math.floor(parseInt(msg.internalDate) / 1000),
            senderName: senderName || undefined,
            email: {
                subject,
                html: htmlBody || undefined,
                to,
                cc: cc.length > 0 ? cc : undefined,
                threadId: msg.threadId,
                messageId,
            },
        };
    }
    extractBody(part, cb) {
        if (!part)
            return;
        if (part.mimeType === 'text/plain' && part.body?.data) {
            cb(Buffer.from(part.body.data, 'base64url').toString('utf-8'));
        }
        if (part.mimeType === 'text/html' && part.body?.data) {
            cb(undefined, Buffer.from(part.body.data, 'base64url').toString('utf-8'));
        }
        if (part.parts) {
            for (const subpart of part.parts) {
                this.extractBody(subpart, cb);
            }
        }
    }
}
exports.GmailChannel = GmailChannel;
