// channelkit/dist/channels/whatsapp/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WhatsAppChannel = void 0;
exports.isBaileysAvailable = isBaileysAvailable;
const path_1 = require("path");
const base_1 = require("../base");
const paths_1 = require("../../paths");
let baileys = null;
async function loadBaileys() {
    if (!baileys) {
        try {
            baileys = await Promise.resolve().then(() => __importStar(require('@whiskeysockets/baileys')));
        }
        catch {
            throw new Error('WhatsApp support requires @whiskeysockets/baileys.\n' +
                'Install it with: npm install @whiskeysockets/baileys\n\n' +
                'Note: @whiskeysockets/baileys is licensed under GPL-3.0.');
        }
    }
    return baileys;
}
function isBaileysAvailable() {
    try {
        require.resolve('@whiskeysockets/baileys');
        return true;
    }
    catch {
        return false;
    }
}
class WhatsAppChannel extends base_1.Channel {
    processedIds = new Set();
    /**
     * Pair a new WhatsApp device by showing a QR code.
     * Creates a temporary Baileys connection, waits for successful pairing, then disconnects.
     * Auth state is persisted to authDir for later use by connect().
     */
    static async pair(authDir, onQR) {
        const { useMultiFileAuthState, fetchLatestBaileysVersion, makeCacheableSignalKeyStore, DisconnectReason, default: makeWASocket } = await loadBaileys();
        const { state, saveCreds } = await useMultiFileAuthState(authDir);
        // Check if already paired (has registered creds)
        const alreadyPaired = state.creds.registered;
        // Silent logger to suppress Baileys noise during pairing
        const silentLogger = {
            level: 'silent',
            child: () => silentLogger,
            trace: () => { },
            debug: () => { },
            info: () => { },
            warn: () => { },
            error: () => { },
            fatal: () => { },
        };
        const { version } = await fetchLatestBaileysVersion();
        const sock = makeWASocket({
            version,
            auth: {
                creds: state.creds,
                keys: makeCacheableSignalKeyStore(state.keys, silentLogger),
            },
            logger: silentLogger,
            browser: ['ChannelKit', 'Desktop', '1.0.0'],
            syncFullHistory: false,
        });
        sock.ev.on('creds.update', saveCreds);
        return new Promise((resolve, reject) => {
            const timeout = setTimeout(() => {
                sock.end(undefined);
                reject(new Error('QR pairing timed out after 60s'));
            }, 60000);
            sock.ev.on('connection.update', (update) => {
                const { connection, lastDisconnect, qr } = update;
                if (qr) {
                    if (onQR)
                        onQR(qr);
                    Promise.resolve().then(() => __importStar(require('qrcode-terminal'))).then(({ default: qrcode }) => {
                        qrcode.generate(qr, { small: true }, (output) => {
                            const indented = output.split('\n').map(line => '     ' + line).join('\n');
                            console.log(indented);
                        });
                    }).catch(() => {
                        console.log(`\n  📱 QR Code: ${qr}\n`);
                    });
                }
                if (connection === 'open') {
                    clearTimeout(timeout);
                    sock.end(undefined);
                    resolve();
                }
                if (connection === 'close') {
                    const reason = lastDisconnect?.error?.output?.statusCode;
                    if (reason === DisconnectReason.loggedOut) {
                        clearTimeout(timeout);
                        sock.end(undefined);
                        reject(new Error('WhatsApp logged out during pairing'));
                    }
                    // If already paired and connection closed normally, that's fine
                    if (alreadyPaired) {
                        clearTimeout(timeout);
                        resolve();
                    }
                }
            });
        });
    }
    sock = null;
    authDir;
    reconnectAttempts = 0;
    reconnectTimer = null;
    static MAX_RECONNECT_ATTEMPTS = 10;
    constructor(name, config, authDir = paths_1.DEFAULT_AUTH_DIR) {
        super(name, config);
        this.connected = false; // WhatsApp channels start disconnected until Baileys connects
        this.authDir = (0, path_1.join)(authDir, `whatsapp-${name}`);
    }
    scheduleReconnect() {
        if (this.reconnectAttempts >= WhatsAppChannel.MAX_RECONNECT_ATTEMPTS) {
            console.error(`[whatsapp:${this.name}] Giving up after ${WhatsAppChannel.MAX_RECONNECT_ATTEMPTS} reconnect attempts`);
            return;
        }
        const delay = Math.min(Math.pow(2, this.reconnectAttempts) * 1000, 30000);
        this.reconnectAttempts++;
        console.log(`[whatsapp:${this.name}] Reconnecting in ${delay / 1000}s (attempt ${this.reconnectAttempts}/${WhatsAppChannel.MAX_RECONNECT_ATTEMPTS})...`);
        this.reconnectTimer = setTimeout(() => this.connect(), delay);
    }
    getSocket() {
        return this.sock;
    }
    async createGroup(name, participants) {
        if (!this.sock)
            throw new Error('WhatsApp not connected');
        const result = await this.sock.groupCreate(name, participants);
        return { id: result.id, subject: name };
    }
    async sendToJid(jid, text) {
        if (!this.sock)
            return;
        await this.sock.sendMessage(jid, { text });
    }
    async connect() {
        this.reconnectAttempts = 0;
        const { useMultiFileAuthState, fetchLatestBaileysVersion, makeCacheableSignalKeyStore, downloadMediaMessage, DisconnectReason, default: makeWASocket } = await loadBaileys();
        const { state, saveCreds } = await useMultiFileAuthState(this.authDir);
        const silentLogger = {
            level: 'silent',
            child: () => silentLogger,
            trace: () => { },
            debug: () => { },
            info: () => { },
            warn: () => { },
            error: () => { },
            fatal: () => { },
        };
        const { version } = await fetchLatestBaileysVersion();
        this.sock = makeWASocket({
            version,
            auth: {
                creds: state.creds,
                keys: makeCacheableSignalKeyStore(state.keys, silentLogger),
            },
            logger: silentLogger,
            browser: ['ChannelKit', 'Desktop', '1.0.0'],
            syncFullHistory: false,
        });
        this.sock.ev.on('creds.update', saveCreds);
        this.sock.ev.on('connection.update', (update) => {
            const { connection, lastDisconnect, qr } = update;
            if (qr) {
                // Emit QR event for dashboard/API
                this.emit('qr', qr);
                // Dynamic import to handle optional dependency
                Promise.resolve().then(() => __importStar(require('qrcode-terminal'))).then(({ default: qrcode }) => {
                    console.log(`\n  📱 Scan this QR code with WhatsApp:\n`);
                    qrcode.generate(qr, { small: true }, (output) => {
                        // Indent the QR code
                        const indented = output.split('\n').map(line => '     ' + line).join('\n');
                        console.log(indented);
                    });
                    console.log(`\n  ${'\x1b[2m'}Open WhatsApp → Settings → Linked Devices → Link a Device${'\x1b[0m'}\n`);
                }).catch(() => {
                    // Fallback: print raw QR string
                    console.log(`\n  📱 QR Code (scan with WhatsApp):\n`);
                    console.log(`  ${qr}\n`);
                    console.log(`  Tip: install qrcode-terminal for a scannable QR in the terminal\n`);
                });
            }
            if (connection === 'close') {
                this.connected = false;
                this.emit('disconnected');
                const reason = lastDisconnect?.error?.output?.statusCode;
                if (reason !== DisconnectReason.loggedOut) {
                    this.scheduleReconnect();
                }
                else {
                    console.log(`[whatsapp:${this.name}] Logged out`);
                }
            }
            else if (connection === 'open') {
                this.connected = true;
                this.reconnectAttempts = 0;
                this.emit('connected');
                console.log(`✅ WhatsApp connected: ${this.name}`);
            }
        });
        this.sock.ev.on('messages.upsert', async ({ messages }) => {
            for (const msg of messages) {
                if (msg.key.fromMe)
                    continue;
                const msgId = msg.key.id;
                if (msgId && this.processedIds.has(msgId))
                    continue;
                if (msgId)
                    this.processedIds.add(msgId);
                // Keep set bounded
                if (this.processedIds.size > 1000) {
                    const arr = [...this.processedIds];
                    this.processedIds = new Set(arr.slice(-500));
                }
                // Resolve LID to regular JID
                if (msg.key.remoteJid?.endsWith('@lid') && msg.key.remoteJidAlt) {
                    msg.key.remoteJid = msg.key.remoteJidAlt;
                }
                const unified = this.toUnified(msg, msg.pushName);
                if (!unified)
                    continue;
                // Download media for messages with media content
                const mediaTypes = ['audio', 'image', 'video', 'document', 'sticker'];
                if (mediaTypes.includes(unified.type) && msg.message) {
                    try {
                        const content = msg.message;
                        let mimetype;
                        let filename;
                        if (content.audioMessage) {
                            mimetype = content.audioMessage.mimetype || 'audio/ogg; codecs=opus';
                        }
                        else if (content.imageMessage) {
                            mimetype = content.imageMessage.mimetype || 'image/jpeg';
                        }
                        else if (content.videoMessage) {
                            mimetype = content.videoMessage.mimetype || 'video/mp4';
                        }
                        else if (content.documentMessage) {
                            mimetype = content.documentMessage.mimetype || 'application/octet-stream';
                            filename = content.documentMessage.fileName || undefined;
                        }
                        else if (content.stickerMessage) {
                            mimetype = content.stickerMessage.mimetype || 'image/webp';
                        }
                        if (mimetype) {
                            const buffer = await downloadMediaMessage(msg, 'buffer', {});
                            console.log(`[whatsapp:${this.name}] Downloaded ${unified.type}: ${buffer.length} bytes, ${mimetype}`);
                            unified.media = { buffer: buffer, mimetype, filename };
                        }
                    }
                    catch (err) {
                        console.error(`[whatsapp:${this.name}] Failed to download ${unified.type} media:`, err);
                    }
                }
                this.emitMessage(unified);
            }
        });
    }
    async disconnect() {
        this.connected = false;
        if (this.reconnectTimer) {
            clearTimeout(this.reconnectTimer);
            this.reconnectTimer = null;
        }
        this.reconnectAttempts = WhatsAppChannel.MAX_RECONNECT_ATTEMPTS; // prevent further reconnects
        this.sock?.end(undefined);
        this.sock = null;
    }
    async send(to, response) {
        if (!this.sock || !this.connected) {
            throw new Error('WhatsApp channel is not connected');
        }
        // Voice message from buffer (TTS output)
        if (response.media?.buffer && response.media.mimetype?.includes('audio')) {
            const mime = response.media.mimetype;
            // WhatsApp voice notes (ptt) require OGG/Opus — other formats (e.g. MP3) are sent as audio files
            const isOggOpus = mime.includes('ogg') || mime.includes('opus');
            await this.sock.sendMessage(to, {
                audio: response.media.buffer,
                mimetype: mime,
                ptt: isOggOpus,
            });
            // If there was also text, don't send it separately (voice replaces text)
            return;
        }
        if (response.text) {
            await this.sock.sendMessage(to, { text: response.text });
        }
        if (response.media?.buffer && response.media.mimetype) {
            // Non-audio buffer (e.g. PDF from webhook response)
            const mime = response.media.mimetype;
            const type = mime.startsWith('image') ? 'image' : 'document';
            await this.sock.sendMessage(to, {
                [type]: response.media.buffer,
                mimetype: mime,
            });
        }
        else if (response.media?.url) {
            const type = response.media.mimetype?.startsWith('image') ? 'image' : 'document';
            await this.sock.sendMessage(to, {
                [type]: { url: response.media.url },
                mimetype: response.media.mimetype,
            });
        }
    }
    toUnified(msg, pushName) {
        const key = msg.key;
        const jid = key.remoteJid;
        if (!jid)
            return null;
        const content = msg.message;
        if (!content)
            return null;
        const text = content.conversation ||
            content.extendedTextMessage?.text ||
            content.imageMessage?.caption ||
            '';
        const isGroup = jid.endsWith('@g.us');
        // Resolve participant: key.participant is primary, msg.participant is fallback
        const participant = key.participant || msg.participant || undefined;
        return {
            id: key.id || `${Date.now()}`,
            channel: 'whatsapp',
            from: isGroup ? (participant || jid) : jid,
            type: content.imageMessage
                ? 'image'
                : content.audioMessage
                    ? 'audio'
                    : content.videoMessage
                        ? 'video'
                        : content.documentMessage
                            ? 'document'
                            : content.stickerMessage
                                ? 'sticker'
                                : 'text',
            text: text || undefined,
            timestamp: msg.messageTimestamp || Math.floor(Date.now() / 1000),
            replyTo: content.extendedTextMessage?.contextInfo?.stanzaId || undefined,
            senderName: pushName || undefined,
            groupId: isGroup ? jid : undefined,
            groupName: undefined,
        };
    }
}
exports.WhatsAppChannel = WhatsAppChannel;
