// channelkit/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelKit = void 0;
const crypto_1 = require("crypto");
const router_1 = require("./core/router");
const server_1 = require("./api/server");
const whatsapp_1 = require("./channels/whatsapp");
const telegram_1 = require("./channels/telegram");
const email_1 = require("./channels/email");
const sms_1 = require("./channels/sms");
const voice_1 = require("./channels/voice");
const endpoint_1 = require("./channels/endpoint");
const onboarding_1 = require("./onboarding");
const logger_1 = require("./core/logger");
const messageHandler_1 = require("./core/messageHandler");
const tunnel_1 = require("./core/tunnel");
const parser_1 = require("./config/parser");
const mcp_1 = require("./mcp");
const updater_1 = require("./core/updater");
const webhook_1 = require("./core/webhook");
class ChannelKit {
    config;
    configPath;
    channels = [];
    channelMap = new Map();
    router;
    apiServer;
    logger;
    onboarding;
    tunnel;
    tunnelStartedBy = null;
    mcpServer;
    updater;
    authSyncTimer;
    constructor(config, configPath) {
        this.config = config;
        this.configPath = configPath;
        // Populate process.env from config.settings (existing env vars take precedence)
        if (config.settings) {
            const map = {
                twilio_account_sid: 'TWILIO_ACCOUNT_SID',
                twilio_auth_token: 'TWILIO_AUTH_TOKEN',
                google_api_key: 'GOOGLE_API_KEY',
                elevenlabs_api_key: 'ELEVENLABS_API_KEY',
                openai_api_key: 'OPENAI_API_KEY',
                deepgram_api_key: 'DEEPGRAM_API_KEY',
                anthropic_api_key: 'ANTHROPIC_API_KEY',
            };
            for (const [key, envVar] of Object.entries(map)) {
                const val = config.settings[key];
                if (val && !process.env[envVar]) {
                    process.env[envVar] = val;
                }
            }
        }
        if (config.settings?.allow_local_webhooks) {
            (0, webhook_1.setAllowLocalWebhooks)(true);
        }
        this.router = new router_1.Router(config.services, config.routes, config.channels);
        if (config.settings)
            this.router.setSettings(config.settings);
        this.apiServer = new server_1.ApiServer(config.apiPort || 4000);
        this.logger = new logger_1.Logger();
        if (config.dashboard?.enabled !== false) {
            this.router.setLogger(this.logger);
            this.apiServer.setLogger(this.logger);
        }
    }
    async start() {
        let whatsappChannel;
        let telegramChannel;
        // Detect duplicate Telegram bot tokens before connecting (causes 409 Conflict)
        const telegramTokensSeen = new Map(); // token → first channel name
        for (const [name, cfg] of Object.entries(this.config.channels)) {
            if (cfg.type === 'telegram') {
                const token = cfg.bot_token;
                if (telegramTokensSeen.has(token)) {
                    console.error(`[channelkit] ERROR: Channels "${telegramTokensSeen.get(token)}" and "${name}" share the same Telegram bot token. ` +
                        `Remove the duplicate from config.yaml or the dashboard before starting.`);
                    process.exit(1);
                }
                telegramTokensSeen.set(token, name);
            }
        }
        // Initialize channels
        for (const [name, channelConfig] of Object.entries(this.config.channels)) {
            let channel;
            switch (channelConfig.type) {
                case 'whatsapp':
                    if (!(0, whatsapp_1.isBaileysAvailable)()) {
                        console.warn(`[channelkit] Skipping WhatsApp channel "${name}" — @whiskeysockets/baileys is not installed.`);
                        console.warn(`[channelkit] Install it with: npm install -g @whiskeysockets/baileys (global) or npm install @whiskeysockets/baileys (local)`);
                        continue;
                    }
                    channel = new whatsapp_1.WhatsAppChannel(name, channelConfig);
                    whatsappChannel = channel;
                    break;
                case 'telegram':
                    channel = new telegram_1.TelegramChannel(name, channelConfig);
                    telegramChannel = channel;
                    break;
                case 'email': {
                    const emailConfig = channelConfig;
                    if (emailConfig.provider === 'gmail') {
                        channel = new email_1.GmailChannel(name, emailConfig);
                    }
                    else if (emailConfig.provider === 'resend') {
                        channel = new email_1.ResendChannel(name, emailConfig);
                    }
                    else {
                        console.warn(`Unknown email provider: ${emailConfig.provider}`);
                        continue;
                    }
                    break;
                }
                case 'sms': {
                    channel = new sms_1.TwilioSMSChannel(name, channelConfig);
                    break;
                }
                case 'voice': {
                    channel = new voice_1.TwilioVoiceChannel(name, channelConfig);
                    break;
                }
                case 'endpoint': {
                    channel = new endpoint_1.EndpointChannel(name, channelConfig);
                    break;
                }
                default:
                    console.warn(`Unknown channel type: ${channelConfig.type}`);
                    continue;
            }
            this.channels.push(channel);
            this.channelMap.set(name, channel);
            // Register by name for API server
            this.apiServer.registerChannel(name, channel);
            // Also register by type for backward compat, but don't overwrite
            // a channel that was registered by its actual name
            if (!this.channelMap.has(channelConfig.type)) {
                this.apiServer.registerChannel(channelConfig.type, channel);
            }
            const mode = this.router.getChannelMode(name);
            console.log(`[channelkit] Channel "${name}" (${channelConfig.type}) → ${mode} mode`);
        }
        // Set up Telegram slash commands for multi-service channels
        for (const [name, channel] of this.channelMap.entries()) {
            if (channel instanceof telegram_1.TelegramChannel) {
                const services = this.router.getServicesForChannel(name);
                if (services.length > 1) {
                    const svcEntries = Object.entries(this.config.services || {})
                        .filter(([_, svc]) => svc.channel === name)
                        .map(([svcName, svc]) => ({ name: svcName, config: svc }));
                    channel.setSlashCommands(svcEntries);
                }
            }
        }
        // Set up onboarding for channels in groups mode
        // Build onboarding codes from services that have codes
        const onboardingCodes = [];
        if (this.config.services) {
            for (const [svcName, svc] of Object.entries(this.config.services)) {
                if (svc.code) {
                    onboardingCodes.push({
                        code: svc.code.toUpperCase(),
                        name: svcName,
                        webhook: svc.webhook,
                        channels: [svc.channel],
                    });
                }
            }
        }
        // Also support legacy onboarding config
        if (this.config.onboarding?.codes) {
            onboardingCodes.push(...this.config.onboarding.codes);
        }
        if (onboardingCodes.length > 0) {
            const onboardingConfig = { codes: onboardingCodes };
            this.onboarding = new onboarding_1.Onboarding(onboardingConfig, whatsappChannel, telegramChannel);
            this.router.setGroupStore(this.onboarding.getGroupStore());
            // Wire up remote store sync if configured
            if (this.config.data_store?.type === 'remote' && this.config.data_store.endpoint) {
                const { RemoteStore } = await Promise.resolve().then(() => __importStar(require('./core/remoteStore')));
                const remoteStore = new RemoteStore(this.config.data_store);
                const groupStore = this.onboarding.getGroupStore();
                // Try to fetch remote groups on startup
                const remoteGroups = await remoteStore.fetchGroups();
                if (remoteGroups) {
                    groupStore.replaceAll(remoteGroups);
                    console.log('[remote-store] Loaded groups from remote endpoint');
                }
                else {
                    console.warn('[remote-store] Remote unreachable, using local groups');
                }
                // Sync changes back to remote
                groupStore.onChange((groups) => remoteStore.pushGroups(groups));
            }
            console.log(`[channelkit] Onboarding enabled with ${onboardingCodes.length} service code(s)`);
        }
        // Wire up remote store for config and auth sync
        if (this.config.data_store?.type === 'remote' && this.config.data_store.endpoint) {
            const { RemoteStore } = await Promise.resolve().then(() => __importStar(require('./core/remoteStore')));
            const remoteStore = new RemoteStore(this.config.data_store);
            // Sync config changes back to remote on every save
            const { setOnSaveHook } = await Promise.resolve().then(() => __importStar(require('./config/parser')));
            setOnSaveHook((yaml) => remoteStore.pushConfig(yaml));
            // Sync auth changes back to remote (watch for file changes)
            this.setupAuthSync(remoteStore);
        }
        // Wire up message handlers
        const handlerDeps = {
            router: this.router,
            apiServer: this.apiServer,
            logger: this.logger,
            onboarding: this.onboarding,
            config: this.config,
        };
        for (const channel of this.channels) {
            (0, messageHandler_1.wireMessageHandler)(channel, handlerDeps);
        }
        // Wire up voice config lookup for API server
        this.apiServer.findVoiceConfig = (channelName) => {
            if (!this.config.services)
                return undefined;
            const svc = Object.values(this.config.services).find(s => s.channel === channelName);
            return svc?.voice;
        };
        // Auto-generate api_secret on first startup if not configured
        if (!this.config.api_secret && this.configPath) {
            const generated = (0, crypto_1.randomBytes)(32).toString('base64url');
            this.config.api_secret = generated;
            try {
                (0, parser_1.saveConfig)(this.configPath, this.config);
                console.log(`\n🔐 New API secret generated and saved to config.yaml:`);
                console.log(`   ${generated}`);
                console.log(`   Use this to log in to the dashboard and for API requests.\n`);
            }
            catch (err) {
                console.warn(`[security] Failed to save generated api_secret: ${err.message}`);
                console.log(`🔐 Generated API secret (not persisted): ${generated}`);
            }
        }
        if (this.config.api_secret) {
            this.apiServer.setApiSecret(this.config.api_secret);
        }
        else {
            console.warn('[security] No api_secret configured — dashboard and API endpoints are unauthenticated.');
        }
        if (this.config.mcp?.secret) {
            this.apiServer.setMcpSecret(this.config.mcp.secret);
        }
        // Start API server + connect all channels
        await this.apiServer.start();
        // If the port changed (user chose a different port due to conflict), save it
        const actualPort = this.apiServer.getPort();
        if (actualPort !== (this.config.apiPort || 4000) && this.configPath) {
            this.config.apiPort = actualPort;
            try {
                (0, parser_1.saveConfig)(this.configPath, this.config);
                console.log(`[config] Port ${actualPort} saved to config.yaml`);
            }
            catch (err) {
                console.warn(`[config] Failed to save new port: ${err.message}`);
            }
        }
        if (this.configPath) {
            this.apiServer.setConfigPath(this.configPath);
        }
        this.apiServer.captureConsole();
        // Wire up hot-reload: when services are changed via dashboard, reload the Router
        this.apiServer.reloadRouter = () => {
            if (!this.configPath)
                return;
            try {
                const freshConfig = (0, parser_1.loadConfig)(this.configPath, { validate: false });
                if (freshConfig.services) {
                    this.router.reloadServices(freshConfig.services, freshConfig.channels);
                    if (freshConfig.settings)
                        this.router.setSettings(freshConfig.settings);
                    console.log('[router] Services reloaded from config');
                }
            }
            catch (err) {
                console.error(`[router] Failed to reload services: ${err.message}`);
            }
        };
        // Set initial external access from config
        if (this.config.tunnel?.expose_dashboard) {
            this.apiServer.setExposeDashboard(true);
        }
        if (this.config.mcp?.expose) {
            this.apiServer.setExposeMcp(true);
        }
        // Wire up tunnel callbacks for dashboard control
        this.apiServer.tunnelStatus = () => ({
            active: this.tunnel?.getPublicUrl() != null,
            url: this.tunnel?.getPublicUrl() || null,
        });
        this.apiServer.tunnelStart = async () => {
            if (this.tunnel?.getPublicUrl()) {
                return { url: this.tunnel.getPublicUrl() };
            }
            const port = this.config.apiPort || 4000;
            // Reload config from file to pick up any saved token/hostname
            let tunnelConfig = { provider: 'cloudflared' };
            if (this.configPath) {
                try {
                    const freshConfig = (0, parser_1.loadConfig)(this.configPath, { validate: false });
                    if (freshConfig.tunnel)
                        tunnelConfig = freshConfig.tunnel;
                }
                catch { }
            }
            this.tunnel = new tunnel_1.TunnelManager(tunnelConfig, port);
            await this.tunnel.start();
            const publicUrl = this.tunnel.getPublicUrl();
            if (!publicUrl)
                throw new Error('Tunnel started but no URL received');
            this.apiServer.setPublicUrl(publicUrl);
            this.tunnelStartedBy = 'dashboard';
            await this.autoConfigureWebhooks(publicUrl);
            return { url: publicUrl };
        };
        this.apiServer.tunnelStop = async () => {
            if (this.tunnel) {
                await this.tunnel.stop();
                this.tunnel = undefined;
                this.apiServer.clearPublicUrl();
                this.tunnelStartedBy = null;
            }
        };
        // Wire up MCP callbacks for dashboard control
        const apiPort = this.config.apiPort || 4000;
        this.apiServer.mcpStatus = () => ({
            active: this.mcpServer != null,
            url: this.mcpServer ? `http://localhost:${apiPort}/mcp` : null,
        });
        this.apiServer.mcpStart = async () => {
            if (this.mcpServer) {
                return { url: `http://localhost:${apiPort}/mcp` };
            }
            const mcpCtx = {
                channels: this.channelMap,
                logger: this.logger,
                configPath: this.configPath,
                config: this.config,
                startTime: Date.now(),
                getPublicUrl: () => this.tunnel?.getPublicUrl() || null,
                updater: this.updater,
            };
            this.mcpServer = new mcp_1.ChannelKitMcpServer(mcpCtx, this.config.mcp || {});
            await this.mcpServer.mountOnExpress(this.apiServer.getExpressApp());
            this.apiServer.setExposeMcp(true);
            return { url: `http://localhost:${apiPort}/mcp` };
        };
        this.apiServer.mcpStop = async () => {
            if (this.mcpServer) {
                await this.mcpServer.stop();
                this.mcpServer = undefined;
                this.apiServer.setExposeMcp(false);
            }
        };
        // Start tunnel if configured and auto_start not explicitly disabled
        if (this.config.tunnel && this.config.tunnel.auto_start !== false) {
            const port = this.config.apiPort || 4000;
            this.tunnel = new tunnel_1.TunnelManager(this.config.tunnel, port);
            try {
                await this.tunnel.start();
                const publicUrl = this.tunnel.getPublicUrl();
                if (publicUrl) {
                    this.apiServer.setPublicUrl(publicUrl);
                    this.tunnelStartedBy = 'cli';
                    await this.autoConfigureWebhooks(publicUrl);
                    // Broadcast so any already-connected dashboard gets the updated state
                    this.apiServer.broadcast({
                        type: 'tunnelStatus',
                        active: true,
                        url: publicUrl,
                        exposeDashboard: !!this.config.tunnel?.expose_dashboard,
                    });
                }
            }
            catch (err) {
                console.error(`[tunnel] Failed to start tunnel: ${err.message}`);
            }
        }
        const connectResults = await Promise.allSettled(this.channels.map((ch) => ch.connect()));
        connectResults.forEach((result, i) => {
            if (result.status === 'rejected') {
                console.error(`[channelkit] Failed to connect channel "${this.channels[i].name}": ${result.reason?.message || result.reason}`);
            }
        });
        console.log('Listening for messages...');
        // Initialize updater
        this.updater = new updater_1.Updater(this.channelMap);
        // Wire up update callbacks for API server
        this.apiServer.updateStatus = async () => {
            return await this.updater.checkForUpdate();
        };
        this.apiServer.updateTrigger = async () => {
            return await this.updater.performUpdate();
        };
        // Start auto-update (enabled by default, but npx always uses notify-only)
        const isNpx = this.updater.getMode() === 'npx';
        if (!isNpx && this.config.auto_update?.enabled !== false) {
            const interval = this.config.auto_update?.interval || 30;
            this.updater.startAutoUpdate(interval);
        }
        else {
            // Auto-update disabled or npx — still check for updates periodically (every 12h) and notify
            this.updater.startUpdateCheck(720);
        }
        // Start MCP server if enabled
        if (this.config.mcp?.enabled) {
            const mcpCtx = {
                channels: this.channelMap,
                logger: this.logger,
                configPath: this.configPath,
                config: this.config,
                startTime: Date.now(),
                getPublicUrl: () => this.tunnel?.getPublicUrl() || null,
                updater: this.updater,
            };
            this.mcpServer = new mcp_1.ChannelKitMcpServer(mcpCtx, this.config.mcp);
            if (this.config.mcp.stdio) {
                // stdio transport — for subprocess usage
                this.mcpServer.startStdio().catch((err) => {
                    console.error(`[mcp] Stdio transport failed: ${err.message}`);
                });
            }
            else {
                // Mount on main Express app (shares port with API server + tunnel)
                await this.mcpServer.mountOnExpress(this.apiServer.getExpressApp());
                // Respect the mcp.expose config (already set earlier at line ~211, but ensure it's set)
                if (this.config.mcp?.expose) {
                    this.apiServer.setExposeMcp(true);
                }
                const publicUrl = this.tunnel?.getPublicUrl();
                if (publicUrl) {
                    console.log(`[mcp] MCP available at ${publicUrl}/mcp (via tunnel)`);
                }
                // Broadcast so any already-connected dashboard gets the updated state
                this.apiServer.broadcast({
                    type: 'mcpStatus',
                    active: true,
                    url: `http://localhost:${apiPort}/mcp`,
                    exposeMcp: !!this.config.mcp?.expose,
                    hasSecret: !!this.config.mcp?.secret,
                });
            }
        }
    }
    async autoConfigureWebhooks(publicUrl) {
        for (const [name, channelConfig] of Object.entries(this.config.channels)) {
            // Auto-configure Twilio webhook for SMS channels
            if (channelConfig.type === 'sms' || channelConfig.account_sid) {
                try {
                    const twilioConfig = channelConfig;
                    if (twilioConfig.account_sid && twilioConfig.auth_token && twilioConfig.number_sid) {
                        const { TwilioProvisioner } = await Promise.resolve().then(() => __importStar(require('./provisioning/twilio')));
                        const twilio = new TwilioProvisioner({
                            accountSid: twilioConfig.account_sid,
                            authToken: twilioConfig.auth_token,
                        });
                        const webhookUrl = `${publicUrl}/inbound/twilio/${name}`;
                        await twilio.client.incomingPhoneNumbers(twilioConfig.number_sid).update({
                            smsUrl: webhookUrl,
                        });
                        console.log(`📱 Updated Twilio webhook for ${twilioConfig.phone_number || name}`);
                    }
                }
                catch (err) {
                    console.error(`[tunnel] Failed to update Twilio webhook for ${name}: ${err.message}`);
                }
            }
            // Auto-configure Resend inbound webhook
            if (channelConfig.type === 'email' && channelConfig.provider === 'resend' && !channelConfig.poll_interval) {
                try {
                    const apiKey = channelConfig.api_key;
                    const webhookUrl = `${publicUrl}/inbound/resend/${name}`;
                    // Check if a webhook already points to our endpoint
                    const listRes = await fetch('https://api.resend.com/webhooks', {
                        headers: { Authorization: `Bearer ${apiKey}` },
                    });
                    let existingId = null;
                    if (listRes.ok) {
                        const list = await listRes.json();
                        const match = list.data?.find((w) => w.endpoint === webhookUrl);
                        if (match)
                            existingId = match.id;
                    }
                    if (!existingId) {
                        // Register new webhook
                        const createRes = await fetch('https://api.resend.com/webhooks', {
                            method: 'POST',
                            headers: { Authorization: `Bearer ${apiKey}`, 'Content-Type': 'application/json' },
                            body: JSON.stringify({ endpoint: webhookUrl, events: ['email.received'] }),
                        });
                        if (createRes.ok) {
                            const result = await createRes.json();
                            channelConfig.webhook_id = result.id;
                            if (result.signing_secret)
                                channelConfig.webhook_secret = result.signing_secret;
                            if (this.configPath)
                                (0, parser_1.saveConfig)(this.configPath, this.config);
                            console.log(`📬 Registered Resend inbound webhook for "${name}" → ${webhookUrl}`);
                        }
                        else {
                            console.error(`[resend] Failed to register webhook: ${await createRes.text()}`);
                            console.log(`📬 Resend inbound webhook (manual): ${webhookUrl}`);
                        }
                    }
                    else {
                        // Already registered — store the ID if missing
                        if (!channelConfig.webhook_id) {
                            channelConfig.webhook_id = existingId;
                            if (this.configPath)
                                (0, parser_1.saveConfig)(this.configPath, this.config);
                        }
                        console.log(`📬 Resend inbound webhook already registered for "${name}" → ${webhookUrl}`);
                    }
                }
                catch (err) {
                    console.error(`[tunnel] Failed to register Resend webhook for ${name}: ${err.message}`);
                }
            }
            // Auto-configure Twilio voice webhook
            if (channelConfig.type === 'voice') {
                const voiceChannel = this.channelMap.get(name);
                if (voiceChannel && voiceChannel instanceof voice_1.TwilioVoiceChannel) {
                    voiceChannel.setPublicUrl(publicUrl);
                    const webhookUrl = `${publicUrl}/inbound/voice/${name}`;
                    await voiceChannel.setWebhookUrl(webhookUrl);
                }
            }
        }
    }
    setupAuthSync(remoteStore) {
        // Sync auth dir to remote every 60 seconds
        const syncAuth = async () => {
            const { DEFAULT_AUTH_DIR } = await Promise.resolve().then(() => __importStar(require('./paths')));
            const { RemoteStore } = await Promise.resolve().then(() => __importStar(require('./core/remoteStore')));
            const zip = await RemoteStore.zipDirectory(DEFAULT_AUTH_DIR);
            if (zip)
                await remoteStore.pushAuth(zip);
        };
        // Initial sync after 10s, then every 60s
        setTimeout(() => {
            syncAuth();
            this.authSyncTimer = setInterval(syncAuth, 60_000);
        }, 10_000);
    }
    async stop() {
        if (this.authSyncTimer) {
            clearInterval(this.authSyncTimer);
        }
        if (this.updater) {
            this.updater.stopAutoUpdate();
        }
        if (this.mcpServer) {
            await this.mcpServer.stop();
        }
        if (this.tunnel) {
            await this.tunnel.stop();
        }
        await this.apiServer.stop();
        await Promise.all(this.channels.map((ch) => ch.disconnect()));
    }
}
exports.ChannelKit = ChannelKit;
