// channelkit/dist/media/processor.js
"use strict";
/**
 * MediaProcessor: handles STT/TTS/AI-formatting for the message pipeline
 *
 * Inbound:  voice message → STT → AI format → adds text to UnifiedMessage
 * Outbound: webhook returns { voice: true } → TTS → sends audio
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.processInbound = processInbound;
exports.processOutbound = processOutbound;
const stt_1 = require("./stt");
const tts_1 = require("./tts");
const formatter_1 = require("./formatter");
const cache = {
    stt: new Map(),
    tts: new Map(),
    format: new Map(),
};
function getSTT(config) {
    if (!config.stt)
        return null;
    const key = `${config.stt.provider}:${config.stt.language || ''}`;
    if (!cache.stt.has(key)) {
        cache.stt.set(key, (0, stt_1.createSTTProvider)(config.stt));
    }
    return cache.stt.get(key);
}
function getTTS(config) {
    if (!config.tts)
        return null;
    const key = `${config.tts.provider}:${config.tts.voice || ''}`;
    if (!cache.tts.has(key)) {
        cache.tts.set(key, (0, tts_1.createTTSProvider)(config.tts));
    }
    return cache.tts.get(key);
}
function getFormatter(config) {
    if (!config.format)
        return null;
    const key = `${config.format.provider}:${config.format.model || 'default'}`;
    if (!cache.format.has(key)) {
        cache.format.set(key, (0, formatter_1.createFormatProvider)(config.format));
    }
    return cache.format.get(key);
}
/**
 * Process inbound message:
 * 1. If audio + STT configured → transcribe to text
 * 2. If format configured → run AI formatting on text
 * Mutates the message in place. Returns format processing info.
 */
async function processInbound(message, serviceConfig) {
    const result = {};
    // Step 1: STT — transcribe audio to text
    if (message.type === 'audio' && serviceConfig.stt && message.media?.buffer) {
        const stt = getSTT(serviceConfig);
        if (stt) {
            try {
                const mimetype = message.media.mimetype || 'audio/ogg';
                console.log(`[stt] Transcribing message ${message.id} (${mimetype}, ${message.media.buffer.length} bytes)...`);
                const transcript = await stt.transcribe(message.media.buffer, mimetype, serviceConfig.stt.language);
                if (transcript) {
                    message.text = transcript;
                    console.log(`[stt] Transcribed: "${transcript.substring(0, 80)}${transcript.length > 80 ? '...' : ''}"`);
                }
                else {
                    console.log(`[stt] Empty transcription for message ${message.id}`);
                }
                // Clear media buffer after transcription unless forward_audio is enabled
                if (!serviceConfig.stt.forward_audio) {
                    delete message.media;
                }
            }
            catch (err) {
                console.error(`[stt] Transcription failed for message ${message.id}:`, err);
            }
        }
    }
    // Step 2: AI formatting — transform text using configured prompt
    if (serviceConfig.format && message.text) {
        const formatter = getFormatter(serviceConfig);
        if (formatter) {
            try {
                console.log(`[format] Formatting message ${message.id} with ${serviceConfig.format.provider}...`);
                const originalText = message.text;
                const formatted = await formatter.format(message.text, serviceConfig.format.prompt);
                if (formatted) {
                    console.log(`[format] Result: "${formatted.substring(0, 80)}${formatted.length > 80 ? '...' : ''}"`);
                    message.text = formatted;
                    result.formatApplied = true;
                    result.formatOriginalText = originalText;
                }
            }
            catch (err) {
                const errMsg = err?.message || String(err);
                console.error(`[format] Formatting failed for message ${message.id}:`, errMsg);
                result.formatError = errMsg.substring(0, 300);
            }
        }
    }
    return result;
}
/** Strip messaging formatting characters so TTS reads clean text. */
function stripFormatting(text) {
    return text
        .replace(/\*([^*]+)\*/g, '$1') // *bold*
        .replace(/_([^_]+)_/g, '$1') // _italic_
        .replace(/~([^~]+)~/g, '$1') // ~strikethrough~
        .replace(/```[\s\S]*?```/g, '') // ```code blocks```
        .replace(/`([^`]+)`/g, '$1') // `inline code`
        .replace(/^>\s?/gm, '') // > blockquote
        .replace(/\[([^\]]+)\]\([^)]+\)/g, '$1') // [link text](url)
        .replace(/!\[([^\]]*)\]\([^)]+\)/g, '$1') // ![alt](image url)
        .replace(/^#{1,6}\s+/gm, '') // # headings
        .replace(/^[-*+]\s+/gm, '') // bullet lists
        .replace(/^\d+\.\s+/gm, '') // numbered lists
        .trim();
}
/**
 * Process outbound response: if TTS is configured on the service, synthesize audio.
 * Returns the response (with media buffer on success) and any TTS error.
 */
async function processOutbound(response, serviceConfig) {
    // Synthesize when TTS is configured on the service, unless the webhook explicitly opted out (voice: false)
    if (response.voice === false || !response.text || !serviceConfig.tts)
        return { response };
    const tts = getTTS(serviceConfig);
    if (!tts)
        return { response };
    try {
        const ttsText = stripFormatting(response.text);
        console.log(`[tts] Synthesizing: "${ttsText.substring(0, 80)}${ttsText.length > 80 ? '...' : ''}"`);
        const { buffer, mimetype } = await tts.synthesize(ttsText, serviceConfig.tts.voice);
        console.log(`[tts] Synthesized ${buffer.length} bytes (${mimetype})`);
        return {
            response: { ...response, media: { buffer, mimetype } },
        };
    }
    catch (err) {
        console.error(`[tts] Synthesis failed:`, err);
        const errMsg = err?.message || String(err);
        return { response, ttsError: errMsg.substring(0, 300) };
    }
}
