// channelkit/dist/api/server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiServer = void 0;
const express_1 = __importDefault(require("express"));
const http_1 = require("http");
const helmet_1 = __importDefault(require("helmet"));
const express_rate_limit_1 = __importDefault(require("express-rate-limit"));
const ws_1 = require("ws");
const crypto_1 = require("crypto");
const auth_1 = require("./middleware/auth");
const send_1 = require("./routes/send");
const inbound_1 = require("./routes/inbound");
const config_1 = require("./routes/config");
const tunnel_1 = require("./routes/tunnel");
const settings_1 = require("./routes/settings");
const twilio_1 = require("./routes/twilio");
const logs_1 = require("./routes/logs");
const dashboard_1 = require("./routes/dashboard");
const restart_1 = require("./routes/restart");
const mcp_1 = require("./routes/mcp");
const update_1 = require("./routes/update");
const export_1 = require("./routes/export");
/** Redact patterns that look like API keys/tokens from log text. */
function redactSecrets(text) {
    // Common API key patterns (long hex, base64, bearer tokens in log output)
    return text
        .replace(/\b(re_[A-Za-z0-9_]{20,})\b/g, 're_****')
        .replace(/\b(sk-[A-Za-z0-9_-]{20,})\b/g, 'sk-****')
        .replace(/\b(whsec_[A-Za-z0-9_]{20,})\b/g, 'whsec_****')
        .replace(/\b(AC[a-f0-9]{32})\b/g, 'AC****')
        .replace(/\b(AIzaSy[A-Za-z0-9_-]{33})\b/g, 'AIza****')
        .replace(/\b(eyJ[A-Za-z0-9_-]{50,})\b/g, 'eyJ****');
}
class ApiServer {
    port;
    app = (0, express_1.default)();
    httpServer = (0, http_1.createServer)(this.app);
    server = null;
    ctx;
    constructor(port = 4000) {
        this.port = port;
        const wss = new ws_1.WebSocketServer({ server: this.httpServer });
        const broadcast = (msg) => {
            const data = JSON.stringify(msg);
            wss.clients.forEach((client) => {
                if (client.readyState === ws_1.WebSocket.OPEN) {
                    client.send(data);
                }
            });
        };
        this.ctx = {
            channels: new Map(),
            logger: undefined,
            configPath: undefined,
            publicUrl: null,
            exposeDashboard: false,
            exposeMcp: false,
            apiSecret: null,
            mcpSecret: null,
            startTime: Date.now(),
            serverLogBuffer: [],
            latestQR: null,
            wss,
            broadcast,
            findVoiceConfig: undefined,
            tunnelStart: undefined,
            tunnelStop: undefined,
            tunnelStatus: undefined,
            mcpStart: undefined,
            mcpStop: undefined,
            mcpStatus: undefined,
            updateStatus: undefined,
            updateTrigger: undefined,
            setPublicUrl: (url) => { this.ctx.publicUrl = url.replace(/\/$/, ''); },
            clearPublicUrl: () => { this.ctx.publicUrl = null; },
            getBaseUrl: () => this.ctx.publicUrl || `http://localhost:${this.port}`,
            getReplyUrl: (channelName, jid) => `${this.ctx.getBaseUrl()}/api/send/${channelName}/${encodeURIComponent(jid)}`,
            setExposeDashboard: (value) => { this.ctx.exposeDashboard = value; },
            setExposeMcp: (value) => { this.ctx.exposeMcp = value; },
        };
        // Security headers
        this.app.use((0, helmet_1.default)({
            contentSecurityPolicy: {
                directives: {
                    defaultSrc: ["'self'"],
                    scriptSrc: ["'self'", "'unsafe-inline'"],
                    styleSrc: ["'self'", "'unsafe-inline'", "https://fonts.googleapis.com"],
                    fontSrc: ["'self'", "https://fonts.gstatic.com"],
                    imgSrc: ["'self'", "data:"],
                    connectSrc: ["'self'", "ws:", "wss:"],
                },
            },
        }));
        // Body size limits
        this.app.use(express_1.default.json({ limit: '1mb' }));
        this.app.use((0, auth_1.mcpCors)(this.ctx));
        this.app.use((0, auth_1.externalAccessGuard)(this.ctx));
        this.app.use((0, auth_1.mcpAuthCheck)(this.ctx));
        this.app.use((0, auth_1.adminAuthCheck)(this.ctx));
        // Rate limiting — stricter for send/inbound, looser for dashboard
        const sendLimiter = (0, express_rate_limit_1.default)({
            windowMs: 60 * 1000,
            max: 60,
            standardHeaders: true,
            legacyHeaders: false,
            message: { error: 'Too many requests, please try again later' },
        });
        const inboundLimiter = (0, express_rate_limit_1.default)({
            windowMs: 60 * 1000,
            max: 120,
            standardHeaders: true,
            legacyHeaders: false,
            message: { error: 'Too many requests' },
        });
        const dashboardLimiter = (0, express_rate_limit_1.default)({
            windowMs: 60 * 1000,
            max: 300,
            standardHeaders: true,
            legacyHeaders: false,
        });
        this.app.use('/api/send', sendLimiter);
        this.app.use('/inbound', inboundLimiter);
        this.app.use('/api', dashboardLimiter);
        // Auth check endpoint — lets the dashboard verify if a secret is required/valid
        this.app.get('/api/auth/check', (_req, res) => {
            if (!this.ctx.apiSecret) {
                res.json({ required: false });
                return;
            }
            const auth = _req.headers.authorization;
            if (auth) {
                const expected = `Bearer ${this.ctx.apiSecret}`;
                const valid = auth.length === expected.length && (0, crypto_1.timingSafeEqual)(Buffer.from(auth), Buffer.from(expected));
                res.json({ required: true, valid });
            }
            else {
                res.json({ required: true, valid: false });
            }
        });
        // WebSocket authentication
        wss.on('connection', (ws, req) => {
            if (this.ctx.apiSecret) {
                const url = new URL(req.url || '/', `http://${req.headers.host || 'localhost'}`);
                const token = url.searchParams.get('token');
                if (!token || token.length !== this.ctx.apiSecret.length ||
                    !(0, crypto_1.timingSafeEqual)(Buffer.from(token), Buffer.from(this.ctx.apiSecret))) {
                    ws.close(4401, 'Unauthorized');
                    return;
                }
            }
        });
        // Register all route modules
        (0, send_1.registerSendRoutes)(this.app, this.ctx);
        (0, inbound_1.registerInboundRoutes)(this.app, this.ctx);
        (0, config_1.registerConfigRoutes)(this.app, this.ctx);
        (0, tunnel_1.registerTunnelRoutes)(this.app, this.ctx);
        (0, settings_1.registerSettingsRoutes)(this.app, this.ctx);
        (0, twilio_1.registerTwilioRoutes)(this.app, this.ctx);
        (0, logs_1.registerLogRoutes)(this.app, this.ctx);
        (0, dashboard_1.registerDashboardRoutes)(this.app, this.ctx);
        (0, restart_1.registerRestartRoutes)(this.app, this.ctx);
        (0, mcp_1.registerMcpRoutes)(this.app, this.ctx);
        (0, update_1.registerUpdateRoutes)(this.app, this.ctx);
        (0, export_1.registerExportRoutes)(this.app, this.ctx);
    }
    // Proxy getters/setters that delegate to ctx for backward compat with index.ts
    setExposeDashboard(value) { this.ctx.setExposeDashboard(value); }
    getExposeDashboard() { return this.ctx.exposeDashboard; }
    setApiSecret(secret) { this.ctx.apiSecret = secret != null ? String(secret) : null; }
    setMcpSecret(secret) { this.ctx.mcpSecret = secret != null ? String(secret) : null; }
    setLogger(logger) {
        this.ctx.logger = logger;
        logger.on('entry', (entry) => {
            this.ctx.broadcast({ type: 'newEntry', entry });
        });
    }
    setConfigPath(path) { this.ctx.configPath = path; }
    captureConsole() {
        const capture = (level, data) => {
            const text = redactSecrets(String(data).replace(/\r?\n$/, ''));
            if (!text)
                return;
            const entry = { level, text, ts: Date.now() };
            this.ctx.serverLogBuffer.push(entry);
            if (this.ctx.serverLogBuffer.length > 500)
                this.ctx.serverLogBuffer.shift();
            this.ctx.broadcast({ type: 'serverLog', ...entry });
        };
        const origOut = process.stdout.write.bind(process.stdout);
        const origErr = process.stderr.write.bind(process.stderr);
        process.stdout.write = (d, ...a) => { capture('stdout', d); return origOut(d, ...a); };
        process.stderr.write = (d, ...a) => { capture('stderr', d); return origErr(d, ...a); };
    }
    registerChannel(name, channel) {
        this.ctx.channels.set(name, channel);
        channel.on('qr', (qr) => {
            this.ctx.latestQR = qr;
            this.ctx.broadcast({ type: 'qr', data: qr });
            // Send raw QR string — frontend renders it
            this.ctx.broadcast({ type: 'whatsapp-qr', channel: name, qr });
        });
        channel.on('connection', () => {
            this.ctx.latestQR = null;
        });
        channel.on('connected', () => {
            this.ctx.broadcast({ type: 'channelStatus', channel: name, connected: true, statusMessage: null });
            this.ctx.broadcast({ type: 'whatsapp-paired', channel: name });
        });
        channel.on('disconnected', () => {
            const statusMessage = channel.statusMessage || null;
            this.ctx.broadcast({ type: 'channelStatus', channel: name, connected: false, statusMessage });
        });
    }
    setPublicUrl(url) { this.ctx.setPublicUrl(url); }
    clearPublicUrl() { this.ctx.clearPublicUrl(); }
    getBaseUrl() { return this.ctx.getBaseUrl(); }
    getReplyUrl(channelName, jid) { return this.ctx.getReplyUrl(channelName, jid); }
    // Direct property setters used by index.ts — proxy to ctx
    set findVoiceConfig(fn) { this.ctx.findVoiceConfig = fn; }
    set tunnelStart(fn) { this.ctx.tunnelStart = fn; }
    set tunnelStop(fn) { this.ctx.tunnelStop = fn; }
    set tunnelStatus(fn) { this.ctx.tunnelStatus = fn; }
    set mcpStart(fn) { this.ctx.mcpStart = fn; }
    set mcpStop(fn) { this.ctx.mcpStop = fn; }
    set mcpStatus(fn) { this.ctx.mcpStatus = fn; }
    set updateStatus(fn) { this.ctx.updateStatus = fn; }
    set updateTrigger(fn) { this.ctx.updateTrigger = fn; }
    set reloadRouter(fn) { this.ctx.reloadRouter = fn; }
    setExposeMcp(value) { this.ctx.setExposeMcp(value); }
    broadcast(msg) { this.ctx.broadcast(msg); }
    getExpressApp() { return this.app; }
    getPort() { return this.port; }
    async start() {
        const portInUse = await this.isPortInUse(this.port);
        if (portInUse) {
            const result = await this.promptPortConflict(this.port);
            if (result === false) {
                console.log('[api] Exiting — port is in use.');
                process.exit(1);
            }
            if (typeof result === 'number') {
                // User chose a different port
                this.port = result;
            }
            this.httpServer = (0, http_1.createServer)(this.app);
            this.ctx.wss = new ws_1.WebSocketServer({ server: this.httpServer });
        }
        if (!this.ctx.apiSecret) {
            console.warn('⚠️  No api_secret configured — dashboard and API endpoints are unprotected. Set api_secret in config.yaml for production use.');
        }
        return new Promise((resolve, reject) => {
            this.server = this.httpServer.listen(this.port, () => {
                console.log(`[api] Async API listening on port ${this.port}`);
                console.log(`📊 Dashboard: http://localhost:${this.port}/dashboard`);
                resolve();
            });
            this.server.on('error', (err) => reject(err));
        });
    }
    isPortInUse(port) {
        return new Promise((resolve) => {
            const tester = require('net').createServer()
                .once('error', (err) => {
                resolve(err.code === 'EADDRINUSE');
            })
                .once('listening', () => {
                tester.close(() => resolve(false));
            })
                .listen(port);
        });
    }
    /** Returns true if process was killed, a number for an alternative port, or false to abort. */
    async promptPortConflict(port) {
        const { execSync } = await Promise.resolve().then(() => __importStar(require('child_process')));
        // Resolve lsof path — macOS keeps it in /usr/sbin which may not be in PATH for background processes
        let lsof = 'lsof';
        try {
            execSync('which lsof', { encoding: 'utf-8', stdio: 'pipe' });
        }
        catch {
            try {
                execSync('/usr/sbin/lsof -v', { encoding: 'utf-8', stdio: 'pipe' });
                lsof = '/usr/sbin/lsof';
            }
            catch { }
        }
        let pid;
        try {
            const out = execSync(`${lsof} -ti tcp:${port}`, { encoding: 'utf-8' }).trim();
            pid = out.split('\n')[0];
        }
        catch { }
        const pidInfo = pid ? ` (PID ${pid})` : '';
        // Detect environments where stdin is unavailable or intercepted:
        // - Not a TTY (piped, CI, etc.)
        // - tsx watch / node --watch: inherits TTY handles so isTTY is true,
        //   but stdin is consumed by the watch runner — readline never receives input.
        const isWatchMode = process.execArgv.some(a => a.includes('--watch'))
            || process.argv.some(a => a === '--watch')
            || !!process.env.TSX_WATCH;
        const autoKill = !process.stdin.isTTY || isWatchMode;
        if (!autoKill) {
            console.error(`\n⚠️  Port ${port} is already in use${pidInfo}.`);
            const { createInterface } = await Promise.resolve().then(() => __importStar(require('readline')));
            const rl = createInterface({ input: process.stdin, output: process.stderr });
            const answer = await new Promise((res) => {
                // Timeout: if readline is broken (e.g. stdin intercepted), auto-kill
                const timeout = setTimeout(() => {
                    rl.close();
                    console.error(`   No response received — auto-killing process on port ${port}...`);
                    res('k');
                }, 10000);
                rl.question(`   [K]ill the existing process, [C]hange port, or [Q]uit? (auto-kill in 10s) `, (ans) => {
                    clearTimeout(timeout);
                    rl.close();
                    res(ans.trim().toLowerCase());
                });
            });
            if (answer === 'c' || answer === 'change') {
                const rl2 = createInterface({ input: process.stdin, output: process.stderr });
                const newPortStr = await new Promise((res) => {
                    rl2.question(`   Enter new port: `, (ans) => {
                        rl2.close();
                        res(ans.trim());
                    });
                });
                const newPort = parseInt(newPortStr, 10);
                if (!newPort || newPort < 1 || newPort > 65535) {
                    console.error(`   Invalid port number.`);
                    return false;
                }
                if (await this.isPortInUse(newPort)) {
                    console.error(`   Port ${newPort} is also in use.`);
                    return false;
                }
                console.log(`   Switching to port ${newPort}.`);
                return newPort;
            }
            if (answer !== 'k' && answer !== 'kill' && answer !== 'y' && answer !== 'yes')
                return false;
        }
        else {
            console.error(`\n⚠️  Port ${port} is already in use${pidInfo}. Auto-killing...`);
        }
        try {
            try {
                execSync(`${lsof} -ti tcp:${port} | xargs kill -9 2>/dev/null`, { encoding: 'utf-8' });
            }
            catch { }
            console.log(`   Killed process on port ${port}. Waiting for port to be released...`);
            for (let i = 0; i < 20; i++) {
                await new Promise((r) => setTimeout(r, 500));
                if (!(await this.isPortInUse(port)))
                    return true;
                try {
                    execSync(`${lsof} -ti tcp:${port} | xargs kill -9 2>/dev/null`, { encoding: 'utf-8' });
                }
                catch { }
            }
            console.error(`   Port ${port} is still in use after waiting.`);
            return false;
        }
        catch (killErr) {
            console.error(`   Failed to kill process: ${killErr.message}`);
            return false;
        }
    }
    async stop() {
        return new Promise((resolve) => {
            this.ctx.wss.close();
            if (this.server) {
                this.server.close(() => resolve());
            }
            else {
                resolve();
            }
        });
    }
}
exports.ApiServer = ApiServer;
