// channelkit/dist/mcp/server.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelKitMcpServer = void 0;
const mcp_js_1 = require("@modelcontextprotocol/sdk/server/mcp.js");
const stdio_js_1 = require("@modelcontextprotocol/sdk/server/stdio.js");
const streamableHttp_js_1 = require("@modelcontextprotocol/sdk/server/streamableHttp.js");
const sse_js_1 = require("@modelcontextprotocol/sdk/server/sse.js");
const types_js_1 = require("@modelcontextprotocol/sdk/types.js");
const crypto_1 = require("crypto");
const express_1 = __importDefault(require("express"));
const http_1 = require("http");
const tools_1 = require("./tools");
/**
 * ChannelKit MCP Server — exposes ChannelKit functionality as MCP tools.
 *
 * Supports three transports:
 * - stdio: for local integrations (Claude Desktop, etc.)
 * - Streamable HTTP (/mcp): for newer clients
 * - SSE (/sse + /messages): for older clients (Antigravity, etc.)
 */
class ChannelKitMcpServer {
    ctx;
    config;
    mcpServer;
    httpApp;
    httpServer;
    httpSessions = new Map();
    sseTransports = new Map();
    constructor(ctx, config = {}) {
        this.ctx = ctx;
        this.config = config;
        this.mcpServer = new mcp_js_1.McpServer({
            name: 'channelkit',
            version: '0.1.0',
        }, {
            capabilities: {
                tools: {},
            },
        });
        (0, tools_1.registerTools)(this.mcpServer, ctx);
    }
    /** Create a fresh McpServer instance with the same tools (for per-connection SSE) */
    createMcpServer() {
        const server = new mcp_js_1.McpServer({ name: 'channelkit', version: '0.1.0' }, { capabilities: { tools: {} } });
        (0, tools_1.registerTools)(server, this.ctx);
        return server;
    }
    /**
     * Start stdio transport (blocks — reads from stdin).
     * Use this when launched as a subprocess by Claude Desktop or similar.
     */
    async startStdio() {
        const transport = new stdio_js_1.StdioServerTransport();
        await this.mcpServer.connect(transport);
        console.error('[mcp] Stdio transport connected');
    }
    /**
     * Start HTTP + SSE transport on the given port.
     * Mounts on /mcp endpoint.
     */
    async startHttp(port) {
        const httpPort = port || this.config.port || 4100;
        this.httpApp = (0, express_1.default)();
        this.httpApp.use(express_1.default.json());
        const transport = new streamableHttp_js_1.StreamableHTTPServerTransport({
            sessionIdGenerator: () => (0, crypto_1.randomUUID)(),
        });
        await this.mcpServer.connect(transport);
        this.httpApp.all('/mcp', async (req, res) => {
            await transport.handleRequest(req, res, req.body);
        });
        // Health check
        this.httpApp.get('/health', (_req, res) => {
            res.json({ status: 'ok', transport: 'streamable-http' });
        });
        this.httpServer = (0, http_1.createServer)(this.httpApp);
        return new Promise((resolve, reject) => {
            this.httpServer.listen(httpPort, () => {
                console.log(`[mcp] HTTP transport listening on port ${httpPort}`);
                console.log(`[mcp] MCP endpoint: http://localhost:${httpPort}/mcp`);
                resolve();
            });
            this.httpServer.on('error', reject);
        });
    }
    /**
     * Mount MCP routes on an existing Express app (shares port with API server).
     * Supports both Streamable HTTP (/mcp) and legacy SSE (/sse + /messages).
     * Each client session gets its own McpServer + transport pair.
     */
    async mountOnExpress(app) {
        // CORS headers + OPTIONS preflight are handled globally by mcpCors() middleware
        // registered in ApiServer before externalAccessGuard and mcpAuthCheck.
        // ── Streamable HTTP transport (newer clients like Claude Desktop) ──
        app.all('/mcp', async (req, res) => {
            try {
                const sessionId = req.headers['mcp-session-id'];
                if (sessionId) {
                    // Existing session — route to its transport
                    const transport = this.httpSessions.get(sessionId);
                    if (transport) {
                        await transport.handleRequest(req, res, req.body);
                    }
                    else {
                        res.status(400).json({
                            jsonrpc: '2.0',
                            error: { code: -32600, message: 'Invalid Request: session not found' },
                            id: null,
                        });
                    }
                    return;
                }
                // New session — only allow POST with initialize request
                if (req.method !== 'POST' || !(0, types_js_1.isInitializeRequest)(req.body)) {
                    res.status(400).json({
                        jsonrpc: '2.0',
                        error: { code: -32600, message: 'Bad Request: first request must be a POST initialize' },
                        id: null,
                    });
                    return;
                }
                const id = (0, crypto_1.randomUUID)();
                const transport = new streamableHttp_js_1.StreamableHTTPServerTransport({
                    sessionIdGenerator: () => id,
                });
                const server = this.createMcpServer();
                await server.connect(transport);
                this.httpSessions.set(id, transport);
                transport.onclose = () => { this.httpSessions.delete(id); };
                await transport.handleRequest(req, res, req.body);
            }
            catch (error) {
                console.error('[mcp] Error handling /mcp request:', error);
                if (!res.headersSent) {
                    res.status(500).json({
                        jsonrpc: '2.0',
                        error: { code: -32603, message: 'Internal server error' },
                        id: null,
                    });
                }
            }
        });
        // ── Legacy SSE transport (Antigravity, older clients) ──
        app.get('/sse', async (req, res) => {
            try {
                const transport = new sse_js_1.SSEServerTransport('/messages', res);
                this.sseTransports.set(transport.sessionId, transport);
                res.on('close', () => {
                    this.sseTransports.delete(transport.sessionId);
                });
                const server = this.createMcpServer();
                await server.connect(transport);
            }
            catch (error) {
                console.error('[mcp] Error handling /sse request:', error);
                if (!res.headersSent) {
                    res.status(500).json({ error: 'Failed to establish SSE connection' });
                }
            }
        });
        app.post('/messages', async (req, res) => {
            try {
                const sessionId = req.query.sessionId;
                const transport = this.sseTransports.get(sessionId);
                if (transport) {
                    await transport.handlePostMessage(req, res, req.body);
                }
                else {
                    res.status(400).json({ error: 'No active SSE session for this sessionId' });
                }
            }
            catch (error) {
                console.error('[mcp] Error handling /messages request:', error);
                if (!res.headersSent) {
                    res.status(500).json({ error: 'Internal server error' });
                }
            }
        });
        console.log('[mcp] MCP endpoints mounted: /mcp (Streamable HTTP), /sse (SSE)');
    }
    async stop() {
        try {
            await this.mcpServer.close();
        }
        catch { }
        if (this.httpServer) {
            return new Promise((resolve) => {
                this.httpServer.close(() => resolve());
            });
        }
    }
}
exports.ChannelKitMcpServer = ChannelKitMcpServer;
