// channelkit/dist/mcp/tools.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerTools = registerTools;
const crypto_1 = require("crypto");
const zod_1 = require("zod");
const restart_1 = require("../core/restart");
const parser_1 = require("../config/parser");
function registerTools(mcp, ctx) {
    // ── Messaging ──────────────────────────────────────────────
    mcp.tool('send_message', 'Send a message through a connected channel', {
        channel: zod_1.z.string().describe('Channel name'),
        to: zod_1.z.string().describe('Recipient identifier. For WhatsApp: phone number (e.g. +972541234567) — automatically converted to JID format. For other channels: chat ID, email, etc.'),
        text: zod_1.z.string().optional().describe('Message text'),
        media: zod_1.z.string().optional().describe('Media URL to attach'),
    }, async ({ channel: channelName, to, text, media }) => {
        const ch = ctx.channels.get(channelName);
        if (!ch) {
            return { content: [{ type: 'text', text: `Channel "${channelName}" not found. Available: ${[...ctx.channels.keys()].join(', ')}` }], isError: true };
        }
        if (!text && !media) {
            return { content: [{ type: 'text', text: 'Must provide text and/or media' }], isError: true };
        }
        // Normalize WhatsApp recipient: convert phone numbers to JID format
        let recipient = to;
        if (ch.config?.type === 'whatsapp' && !to.includes('@')) {
            recipient = to.replace(/[^0-9]/g, '') + '@s.whatsapp.net';
        }
        try {
            await ch.send(recipient, { text, media: media ? { url: media } : undefined });
            if (ctx.logger) {
                ctx.logger.log({
                    id: (0, crypto_1.randomUUID)(),
                    timestamp: Math.floor(Date.now() / 1000),
                    channel: channelName,
                    from: 'system (mcp)',
                    senderName: 'MCP',
                    text: text || '(media)',
                    type: 'mcp-outbound',
                    route: 'mcp:send_message',
                    status: 'success',
                    latency: 0,
                });
            }
            return { content: [{ type: 'text', text: `Message sent via ${channelName} to ${to}` }] };
        }
        catch (err) {
            return { content: [{ type: 'text', text: `Failed to send: ${err.message}` }], isError: true };
        }
    });
    mcp.tool('get_messages', 'Get message history from the logs', {
        channel: zod_1.z.string().optional().describe('Filter by channel name'),
        limit: zod_1.z.number().optional().describe('Max number of messages (default 50)'),
        search: zod_1.z.string().optional().describe('Search text in messages'),
    }, async ({ channel, limit, search }) => {
        if (!ctx.logger) {
            return { content: [{ type: 'text', text: 'Logger not available' }], isError: true };
        }
        const results = ctx.logger.search({
            limit: limit || 50,
            channel: channel || undefined,
            search: search || undefined,
        });
        return { content: [{ type: 'text', text: JSON.stringify(results, null, 2) }] };
    });
    // ── Channel Management ─────────────────────────────────────
    mcp.tool('list_channels', 'List all configured channels with their status', {}, async () => {
        const channels = {};
        if (ctx.configPath) {
            try {
                const config = (0, parser_1.loadConfig)(ctx.configPath, { validate: false });
                for (const [name, cfg] of Object.entries(config.channels)) {
                    channels[name] = {
                        type: cfg.type,
                        mode: cfg.mode,
                        connected: ctx.channels.has(name),
                    };
                }
            }
            catch {
                // Fall back to runtime channels
                for (const [name, ch] of ctx.channels) {
                    channels[name] = { type: ch.config?.type, connected: true };
                }
            }
        }
        else {
            for (const [name, ch] of ctx.channels) {
                channels[name] = { type: ch.config?.type, connected: true };
            }
        }
        return { content: [{ type: 'text', text: JSON.stringify(channels, null, 2) }] };
    });
    mcp.tool('add_channel', 'Add a new channel to the configuration. Requires restart to take effect — set auto_restart to restart automatically.', {
        name: zod_1.z.string().describe('Channel name'),
        type: zod_1.z.enum(['whatsapp', 'telegram', 'email', 'sms', 'voice']).describe('Channel type'),
        config: zod_1.z.record(zod_1.z.string(), zod_1.z.any()).optional().describe('Channel-specific configuration (bot_token, api_key, etc.)'),
        allow_list: zod_1.z.array(zod_1.z.string()).optional().describe('Optional allow list of sender identifiers (phone numbers). If set, only these senders can use the channel.'),
        auto_restart: zod_1.z.boolean().optional().default(false).describe('Automatically restart ChannelKit to apply changes'),
    }, async ({ name, type, config: channelConfig, allow_list, auto_restart }) => {
        if (!ctx.configPath) {
            return { content: [{ type: 'text', text: 'Config path not set — cannot modify config' }], isError: true };
        }
        try {
            const config = (0, parser_1.loadConfig)(ctx.configPath, { validate: false });
            if (config.channels[name]) {
                return { content: [{ type: 'text', text: `Channel "${name}" already exists` }], isError: true };
            }
            config.channels[name] = {
                type,
                ...channelConfig,
                ...(allow_list && allow_list.length > 0 && { allow_list }),
            };
            (0, parser_1.saveConfig)(ctx.configPath, config);
            if (auto_restart) {
                setTimeout(() => (0, restart_1.restartProcess)(ctx.channels), 500);
                return { content: [{ type: 'text', text: `Channel "${name}" (${type}) added. ChannelKit is restarting...` }] };
            }
            return { content: [{ type: 'text', text: `Channel "${name}" (${type}) added. Restart ChannelKit to connect it.` }] };
        }
        catch (err) {
            return { content: [{ type: 'text', text: `Failed: ${err.message}` }], isError: true };
        }
    });
    mcp.tool('remove_channel', 'Remove a channel and its dependent services from the configuration. Requires restart to apply — set auto_restart to restart automatically.', {
        name: zod_1.z.string().describe('Channel name to remove'),
        auto_restart: zod_1.z.boolean().optional().default(false).describe('Automatically restart ChannelKit to apply changes'),
    }, async ({ name, auto_restart }) => {
        if (!ctx.configPath) {
            return { content: [{ type: 'text', text: 'Config path not set' }], isError: true };
        }
        try {
            const config = (0, parser_1.loadConfig)(ctx.configPath, { validate: false });
            if (!config.channels[name]) {
                return { content: [{ type: 'text', text: `Channel "${name}" not found` }], isError: true };
            }
            delete config.channels[name];
            const removedServices = [];
            if (config.services) {
                for (const [svcName, svc] of Object.entries(config.services)) {
                    if (svc.channel === name) {
                        delete config.services[svcName];
                        removedServices.push(svcName);
                    }
                }
            }
            (0, parser_1.saveConfig)(ctx.configPath, config);
            let msg = `Channel "${name}" removed.`;
            if (removedServices.length > 0)
                msg += ` Also removed services: ${removedServices.join(', ')}.`;
            if (auto_restart) {
                setTimeout(() => (0, restart_1.restartProcess)(ctx.channels), 500);
                msg += ' ChannelKit is restarting...';
            }
            else {
                msg += ' Restart ChannelKit to apply.';
            }
            return { content: [{ type: 'text', text: msg }] };
        }
        catch (err) {
            return { content: [{ type: 'text', text: `Failed: ${err.message}` }], isError: true };
        }
    });
    mcp.tool('set_channel_mode', 'Set a channel\'s routing mode to "service" (single service, direct routing) or "groups" (multiple services via codes/commands). Requires restart to apply — set auto_restart to restart automatically.', {
        name: zod_1.z.string().describe('Channel name'),
        mode: zod_1.z.enum(['service', 'groups']).describe('Routing mode'),
        auto_restart: zod_1.z.boolean().optional().default(false).describe('Automatically restart ChannelKit to apply changes'),
    }, async ({ name, mode, auto_restart }) => {
        if (!ctx.configPath) {
            return { content: [{ type: 'text', text: 'Config path not set — cannot modify config' }], isError: true };
        }
        try {
            const config = (0, parser_1.loadConfig)(ctx.configPath, { validate: false });
            if (!config.channels[name]) {
                return { content: [{ type: 'text', text: `Channel "${name}" not found` }], isError: true };
            }
            config.channels[name].mode = mode;
            (0, parser_1.saveConfig)(ctx.configPath, config);
            if (auto_restart) {
                setTimeout(() => (0, restart_1.restartProcess)(ctx.channels), 500);
                return { content: [{ type: 'text', text: `Channel "${name}" mode set to "${mode}". ChannelKit is restarting...` }] };
            }
            return { content: [{ type: 'text', text: `Channel "${name}" mode set to "${mode}". Restart ChannelKit to apply.` }] };
        }
        catch (err) {
            return { content: [{ type: 'text', text: `Failed: ${err.message}` }], isError: true };
        }
    });
    // ── Service Management ─────────────────────────────────────
    mcp.tool('list_services', 'List all configured services', {}, async () => {
        if (!ctx.configPath) {
            return { content: [{ type: 'text', text: JSON.stringify(ctx.config.services || {}, null, 2) }] };
        }
        try {
            const config = (0, parser_1.loadConfig)(ctx.configPath, { validate: false });
            return { content: [{ type: 'text', text: JSON.stringify(config.services || {}, null, 2) }] };
        }
        catch (err) {
            return { content: [{ type: 'text', text: `Failed: ${err.message}` }], isError: true };
        }
    });
    mcp.tool('add_service', 'Create a new service. Requires restart to activate — set auto_restart to restart automatically.', {
        name: zod_1.z.string().describe('Service name'),
        channel: zod_1.z.string().describe('Channel name this service uses'),
        webhook: zod_1.z.string().describe('Webhook URL for this service'),
        code: zod_1.z.string().optional().describe('Magic code for onboarding (groups mode)'),
        command: zod_1.z.string().optional().describe('Slash command for Telegram multi-service'),
        stt: zod_1.z.object({
            provider: zod_1.z.enum(['google', 'whisper', 'deepgram']),
            language: zod_1.z.string().optional(),
            forward_audio: zod_1.z.boolean().optional().describe('Send original audio to webhook along with transcription (default: false)'),
        }).optional().describe('Speech-to-text config'),
        tts: zod_1.z.object({
            provider: zod_1.z.enum(['google', 'elevenlabs', 'openai']),
            voice: zod_1.z.string().optional(),
            language: zod_1.z.string().optional(),
        }).optional().describe('Text-to-speech config'),
        allow_list: zod_1.z.array(zod_1.z.string()).optional().describe('Optional allow list of sender identifiers (phone numbers). If set, only these senders can use the service.'),
        auto_restart: zod_1.z.boolean().optional().default(false).describe('Automatically restart ChannelKit to apply changes'),
    }, async ({ name, channel, webhook, code, command, stt, tts, allow_list, auto_restart }) => {
        if (!ctx.configPath) {
            return { content: [{ type: 'text', text: 'Config path not set' }], isError: true };
        }
        try {
            const config = (0, parser_1.loadConfig)(ctx.configPath, { validate: false });
            if (!config.services)
                config.services = {};
            if (config.services[name]) {
                return { content: [{ type: 'text', text: `Service "${name}" already exists` }], isError: true };
            }
            if (!config.channels[channel]) {
                return { content: [{ type: 'text', text: `Channel "${channel}" does not exist` }], isError: true };
            }
            const svc = { channel, webhook };
            if (code)
                svc.code = code;
            if (command)
                svc.command = command;
            if (stt)
                svc.stt = stt;
            if (tts)
                svc.tts = tts;
            if (allow_list && allow_list.length > 0)
                svc.allow_list = allow_list;
            config.services[name] = svc;
            (0, parser_1.saveConfig)(ctx.configPath, config);
            if (auto_restart) {
                setTimeout(() => (0, restart_1.restartProcess)(ctx.channels), 500);
                return { content: [{ type: 'text', text: `Service "${name}" created. ChannelKit is restarting...` }] };
            }
            return { content: [{ type: 'text', text: `Service "${name}" created. Restart ChannelKit to activate.` }] };
        }
        catch (err) {
            return { content: [{ type: 'text', text: `Failed: ${err.message}` }], isError: true };
        }
    });
    mcp.tool('update_service', 'Update an existing service. Requires restart to apply — set auto_restart to restart automatically.', {
        name: zod_1.z.string().describe('Service name to update'),
        webhook: zod_1.z.string().optional().describe('New webhook URL'),
        code: zod_1.z.string().optional().describe('New magic code'),
        command: zod_1.z.string().optional().describe('New slash command'),
        stt: zod_1.z.object({
            provider: zod_1.z.enum(['google', 'whisper', 'deepgram']),
            language: zod_1.z.string().optional(),
            forward_audio: zod_1.z.boolean().optional().describe('Send original audio to webhook along with transcription (default: false)'),
        }).optional().describe('New STT config'),
        tts: zod_1.z.object({
            provider: zod_1.z.enum(['google', 'elevenlabs', 'openai']),
            voice: zod_1.z.string().optional(),
            language: zod_1.z.string().optional(),
        }).optional().describe('New TTS config'),
        allow_list: zod_1.z.array(zod_1.z.string()).optional().describe('Allow list of sender identifiers. Pass empty array to remove restrictions.'),
        auto_restart: zod_1.z.boolean().optional().default(false).describe('Automatically restart ChannelKit to apply changes'),
    }, async ({ name, webhook, code, command, stt, tts, allow_list, auto_restart }) => {
        if (!ctx.configPath) {
            return { content: [{ type: 'text', text: 'Config path not set' }], isError: true };
        }
        try {
            const config = (0, parser_1.loadConfig)(ctx.configPath, { validate: false });
            if (!config.services?.[name]) {
                return { content: [{ type: 'text', text: `Service "${name}" not found` }], isError: true };
            }
            if (webhook)
                config.services[name].webhook = webhook;
            if (code !== undefined) {
                if (code)
                    config.services[name].code = code;
                else
                    delete config.services[name].code;
            }
            if (command !== undefined) {
                if (command)
                    config.services[name].command = command;
                else
                    delete config.services[name].command;
            }
            if (stt !== undefined) {
                if (stt)
                    config.services[name].stt = stt;
                else
                    delete config.services[name].stt;
            }
            if (tts !== undefined) {
                if (tts)
                    config.services[name].tts = tts;
                else
                    delete config.services[name].tts;
            }
            if (allow_list !== undefined) {
                if (allow_list.length > 0)
                    config.services[name].allow_list = allow_list;
                else
                    delete config.services[name].allow_list;
            }
            (0, parser_1.saveConfig)(ctx.configPath, config);
            if (auto_restart) {
                setTimeout(() => (0, restart_1.restartProcess)(ctx.channels), 500);
                return { content: [{ type: 'text', text: `Service "${name}" updated. ChannelKit is restarting...` }] };
            }
            return { content: [{ type: 'text', text: `Service "${name}" updated. Restart to apply changes.` }] };
        }
        catch (err) {
            return { content: [{ type: 'text', text: `Failed: ${err.message}` }], isError: true };
        }
    });
    mcp.tool('remove_service', 'Remove a service from the configuration. Requires restart to apply — set auto_restart to restart automatically.', {
        name: zod_1.z.string().describe('Service name to remove'),
        auto_restart: zod_1.z.boolean().optional().default(false).describe('Automatically restart ChannelKit to apply changes'),
    }, async ({ name, auto_restart }) => {
        if (!ctx.configPath) {
            return { content: [{ type: 'text', text: 'Config path not set' }], isError: true };
        }
        try {
            const config = (0, parser_1.loadConfig)(ctx.configPath, { validate: false });
            if (!config.services?.[name]) {
                return { content: [{ type: 'text', text: `Service "${name}" not found` }], isError: true };
            }
            delete config.services[name];
            (0, parser_1.saveConfig)(ctx.configPath, config);
            if (auto_restart) {
                setTimeout(() => (0, restart_1.restartProcess)(ctx.channels), 500);
                return { content: [{ type: 'text', text: `Service "${name}" removed. ChannelKit is restarting...` }] };
            }
            return { content: [{ type: 'text', text: `Service "${name}" removed. Restart to apply.` }] };
        }
        catch (err) {
            return { content: [{ type: 'text', text: `Failed: ${err.message}` }], isError: true };
        }
    });
    // ── Status ─────────────────────────────────────────────────
    mcp.tool('get_status', 'Get ChannelKit status: connected channels, uptime, message stats, current version, and update availability', {}, async () => {
        const channelStatus = {};
        for (const [name, ch] of ctx.channels) {
            channelStatus[name] = {
                type: ch.config?.type,
                connected: true,
            };
        }
        const stats = ctx.logger?.getStats() || { total: 0, byChannel: {}, avgLatency: 0, errorCount: 0 };
        const uptime = Date.now() - ctx.startTime;
        // Version info
        let version = {};
        if (ctx.updater) {
            try {
                const updateStatus = await ctx.updater.checkForUpdate();
                version = {
                    mode: updateStatus.mode,
                    current_version: updateStatus.currentVersion,
                    latest_version: updateStatus.latestVersion,
                    update_available: updateStatus.updateAvailable,
                    behind_count: updateStatus.behindCount,
                };
            }
            catch {
                version = { current_version: ctx.updater.getCurrentVersion(), update_available: false };
            }
        }
        else {
            try {
                const { execSync } = await Promise.resolve().then(() => __importStar(require('child_process')));
                version.current_version = execSync('git rev-parse --short HEAD', { cwd: process.cwd(), encoding: 'utf-8' }).trim();
            }
            catch {
                version.current_version = 'unknown';
            }
        }
        const status = {
            uptime_ms: uptime,
            uptime_human: formatUptime(uptime),
            channels: channelStatus,
            messages: stats,
            public_url: ctx.getPublicUrl(),
            version,
        };
        return { content: [{ type: 'text', text: JSON.stringify(status, null, 2) }] };
    });
    mcp.tool('update', 'Update ChannelKit to the latest version. Pulls latest changes (git) or installs latest package (npm), and restarts the process. The MCP connection will be lost after restart.', {}, async () => {
        if (!ctx.updater) {
            return { content: [{ type: 'text', text: 'Updater not available' }], isError: true };
        }
        try {
            const result = await ctx.updater.performUpdate();
            if (result.success) {
                return { content: [{ type: 'text', text: `Update successful: ${result.previousVersion} -> ${result.newVersion}. ChannelKit is restarting... The MCP connection will reconnect shortly.` }] };
            }
            return { content: [{ type: 'text', text: `Update failed: ${result.error}` }], isError: true };
        }
        catch (err) {
            return { content: [{ type: 'text', text: `Update failed: ${err.message}` }], isError: true };
        }
    });
    // ── Restart ──────────────────────────────────────────────
    mcp.tool('restart', 'Restart the ChannelKit process. The MCP connection will be lost and should reconnect automatically.', {}, async () => {
        setTimeout(() => (0, restart_1.restartProcess)(ctx.channels), 500);
        return { content: [{ type: 'text', text: 'ChannelKit is restarting... The MCP connection will reconnect shortly.' }] };
    });
    // ── Configuration ───────────────────────────────────────────
    mcp.tool('set_config', 'Set a configuration value using a dot-separated path. Examples: set_config("apiPort", 5000), set_config("tunnel.expose_dashboard", true), set_config("settings.openai_api_key", "sk-..."), set_config("mcp.enabled", true). Changes are saved to config.yaml. Some changes require a restart to take effect.', {
        path: zod_1.z.string().describe('Dot-separated config path (e.g. "apiPort", "tunnel.token", "settings.openai_api_key", "mcp.enabled")'),
        value: zod_1.z.any().describe('Value to set (string, number, boolean, or null to delete the key)'),
        auto_restart: zod_1.z.boolean().optional().default(false).describe('Automatically restart ChannelKit to apply changes'),
    }, async ({ path, value, auto_restart }) => {
        if (!ctx.configPath) {
            return { content: [{ type: 'text', text: 'Config path not set — cannot modify config' }], isError: true };
        }
        try {
            const config = (0, parser_1.loadConfig)(ctx.configPath, { validate: false });
            const parts = path.split('.');
            let target = config;
            // Navigate to parent
            for (let i = 0; i < parts.length - 1; i++) {
                if (target[parts[i]] === undefined || target[parts[i]] === null) {
                    target[parts[i]] = {};
                }
                target = target[parts[i]];
            }
            const key = parts[parts.length - 1];
            if (value === null) {
                delete target[key];
            }
            else {
                target[key] = value;
            }
            (0, parser_1.saveConfig)(ctx.configPath, config);
            const display = value === null ? '(deleted)' : JSON.stringify(value);
            if (auto_restart) {
                setTimeout(() => (0, restart_1.restartProcess)(ctx.channels), 500);
                return { content: [{ type: 'text', text: `Set ${path} = ${display}. ChannelKit is restarting...` }] };
            }
            return { content: [{ type: 'text', text: `Set ${path} = ${display}. Restart ChannelKit if needed to apply.` }] };
        }
        catch (err) {
            return { content: [{ type: 'text', text: `Failed: ${err.message}` }], isError: true };
        }
    });
}
function formatUptime(ms) {
    const s = Math.floor(ms / 1000);
    const d = Math.floor(s / 86400);
    const h = Math.floor((s % 86400) / 3600);
    const m = Math.floor((s % 3600) / 60);
    const parts = [];
    if (d > 0)
        parts.push(`${d}d`);
    if (h > 0)
        parts.push(`${h}h`);
    parts.push(`${m}m`);
    return parts.join(' ');
}
